åœ¨ubuntué‡Œé¢æ‰“å¼€
# 1. å‰æœŸå‡†å¤‡

```
# å®šä½ 
cd /mnt/c/shengxin 

- æ–°å»ºä¸€ä¸ªbiosoftæ–‡ä»¶å¤¹ç”¨äºå­˜æ”¾ç”Ÿç‰©è½¯ä»¶å·¥å…·
mkdir biosoft

- æ–°å»ºä¸€ä¸ªé¡¹ç›®æ–‡ä»¶å¤¹å…¶ä¸­åŒ…å«å¤§é¼ çš„æ–‡ä»¶å¤¹
mkdir -p project/rat
 mkdirç”¨æ³•ï¼š
mkdir [-p] dirName,-p ç¡®ä¿ç›®å½•åç§°å­˜åœ¨ï¼Œä¸å­˜åœ¨çš„å°±å»ºä¸€ä¸ª

- åœ¨å·¥ä½œç›®å½•ä¸‹çš„projectç›®å½•ä¸­ï¼Œå»ºç«‹ä¸€ä¸ªåä¸º ratçš„å­ç›®å½•

# è¿›å…¥
cd /mnt/c/shengxin/project/rat

# æ–°å»ºå„ç±»å­˜æ”¾çš„æ–‡ä»¶å¤¹
mkdir annotation genome sequence output script

# è½¬åˆ°cd project/ratæ–‡ä»¶å¤¹ä¸‹çœ‹å·²å»ºç«‹çš„ç›®å½•ç»“æ„
cd /mnt/c/shengxin/project/rat
tree
```
ç»“æœ
```
.
â”œâ”€â”€ annotation  ç”¨äºå­˜æ”¾å¤§é¼ çš„åŸºå› ç»„æ³¨é‡Šä¿¡æ¯(.gff/gtf)
â”œâ”€â”€ genome      ç”¨äºå­˜æ”¾å¤§é¼ çš„åŸºå› ç»„æ•°æ®(.fasta)
â”œâ”€â”€ output      ç”¨äºå­˜æ”¾å¤„ç†åçš„æ•°æ®
â”œâ”€â”€ script      ç”¨äºéƒ¨åˆ†è„šæœ¬
â””â”€â”€ sequence    ç”¨äºå­˜æ”¾æµ‹åºåŸå§‹æ•°æ®

5 directories, 0 files
```
# 2. å·¥å…·

2.0 condaå¯ä»¥å¾ˆæ–¹ä¾¿çš„å®‰è£…å’Œç®¡ç†ç”Ÿä¿¡ç›¸å…³çš„å·¥å…·

## 2.1 sratoolkit
sraæ˜¯NCBIçš„ç”¨äºä¸‹è½½æ•°æ®ä»¥åŠè½¬åŒ–æ•°æ®ä½¿ç”¨

- brewå®‰è£…
```
brew install sratoolkit

# å¯¼å…¥ä¸´æ—¶ç¯å¢ƒå˜é‡
$ export PATH="$(pwd):$PATH"
$ prefetch --help
```

## 2.2 fastqc
å¯¹æµ‹åºæ–‡ä»¶è´¨é‡æ§åˆ¶

```
brew install fastqc
```
## 2.3 multiqc
å°†fastqcçš„ç»Ÿè®¡ç»“æœæ±‡èšæˆä¸€ä¸ªç½‘é¡µå¯è§†åŒ–æ–‡ä»¶ï¼Œä¾¿äºæŸ¥çœ‹

- ä½¿ç”¨pythonçš„å®‰è£…å™¨å®‰è£…
```
pip install multiqc
```
## 2.4 cutadapt
ç”¨äºå»é™¤æµ‹åºæ¥å¤´
```
pip install cutadapt
```
## 2.5 Trim Galore
ä½¿ç”¨perlè„šæœ¬ç¼–å†™çš„å·¥å…·ï¼Œæ˜¯å¯¹cutapaterå’Œfastqcå‘½ä»¤çš„å°è£…ã€‚æ˜¯è‡ªåŠ¨æ£€æµ‹adapterçš„è´¨æ§è½¯ä»¶ã€‚é€‚ç”¨äºæ‰€æœ‰é«˜é€šé‡æµ‹åºï¼ŒåŒ…æ‹¬RRBS(Reduced Representation Bisulfite-Seq ), Illuminaã€Nextera å’Œsmall RNAæµ‹åºå¹³å°çš„åŒç«¯å’Œå•ç«¯æ•°æ®

```
cd /mnt/c/shengxin/biosoft

wget https://github.com/FelixKrueger/TrimGalore/archive/0.6.3.tar.gz -O TrimGalore.gz

gzip -d 0.6.10.tar.gz
#gzip -d --decompress --uncompress è§£å‹
```

- fastp: ä¸€æ¬¾è¶…å¿«é€Ÿå…¨åŠŸèƒ½çš„FASTQæ–‡ä»¶è‡ªåŠ¨åŒ–è´¨æ§+è¿‡æ»¤+æ ¡æ­£+é¢„å¤„ç†è½¯ä»¶
- trimmomatic

trimmomaticæ˜¯ä¸€æ¬¾å¤šçº¿ç¨‹å‘½ä»¤è¡Œå·¥å…·ï¼Œå¯ä»¥ç”¨æ¥ä¿®å‰ªIllumina (FASTQ)æ•°æ®ä»¥åŠåˆ é™¤æ¥å¤´ï¼Œæ˜¯ç›®å‰ä½¿ç”¨æœ€å¤šçš„é«˜é€šé‡æµ‹åºæ•°æ®æ¸…æ´—çš„å·¥å…·ã€‚

å®‰è£…å¦‚ä¸‹
```
cd /mnt/c/shengxin/biosoft
wget http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.38.zip
unzip Trimmomatic-0.38.zip

cd Trimmomatic-0.38
export PATH="$(pwd):$PATH"ï¼ˆå¯¼å…¥ä¸´æ—¶ç¯å¢ƒå˜é‡ï¼‰
```
## 2.6 hisat2
Hisat2ä¸»è¦ç”¨äºè½¬å½•ç»„æ•°æ®çš„æ¯”å¯¹

é¦–å…ˆç”¨æµè§ˆå™¨è¿›å…¥[hisat2](https://daehwankimlab.github.io/hisat2/)
åœ¨å³ä¾§æœ‰å¯¹åº”çš„ä¸åŒç³»ç»Ÿçš„å®‰è£…ç‰ˆæœ¬ï¼Œæ ¹æ®è‡ªå·±çš„ç³»ç»Ÿè¿›è¡Œé€‰æ‹©

```
cd /mnt/c/shengxin/biosoft
wget https://cloud.biohpc.swmed.edu/index.php/s/hisat2-210-Linux_x86_64/download
unzip download
cd hisat2-2.1.0

- å¯¼å…¥ä¸´æ—¶ç¯å¢ƒå˜é‡
export PATH="~/biosoft/hisat2-2.1.0:$PATH"

- æµ‹è¯•æ˜¯å¦å¯ç”¨
hisat2 -h

#å‰é¢éƒ½æ²¡æœ‰é—®é¢˜ï¼Œä½†æ˜¯åœ¨è¿™é‡Œä¸€ç›´å‡ºç°
Command 'hisat2' not found, but can be installed with:

sudo apt install hisat2
#è¿è¡Œsudo apt install hisat2å‡ºç°E: Unable to locate package hisat2
è§£å†³æ–¹æ¡ˆ sudo apt update#æ›´æ–°å­˜å‚¨åº“ç¼“å­˜
```
## 2.7 samtools
æ¯”å¯¹å¾—åˆ°çš„samæˆ–è€…bamæ–‡ä»¶çš„å„ç§æ“ä½œ

 SAM (Sequence Alignment/Map)æ ¼å¼æ˜¯ä¸€ç§é€šç”¨çš„æ¯”å¯¹æ ¼å¼ï¼Œç”¨æ¥å­˜å‚¨readsåˆ°å‚è€ƒåºåˆ—çš„æ¯”å¯¹ä¿¡æ¯

 bam æ˜¯sam åˆ™æ˜¯ä¸€ç§æ¯”gzæ›´åŠ é«˜æ•ˆçš„å‹ç¼©ç®—æ³•

```
brew install samtoolsä¸€ç›´æŠ¥é”™
```
æ¢æºä»£ç å®‰è£…
```
cd /mnt/c/shengxin/biosoft
wget -c https://github.com/samtools/samtools/releases/download/1.17/samtools-1.17.tar.bz2 #wget -c æ–­ç‚¹ç»­ä¼ ,å¯¹äºå¤§æ–‡ä»¶ä¸‹è½½æ—¶éœ€è¦ç»§ç»­ä¸­æ–­çš„ä¸‹è½½æ—¶å¯ä»¥ä½¿ç”¨-cå‚æ•°)
tar jxvf samtools-1.17.tar.bz2 #tar jxvfæ˜¯è§£å‹æŒ‡ä»¤ï¼ŒæŒ‡è§£å‹bzip2æ ¼å¼çš„æ–‡ä»¶ï¼Œj-è§£å‹æŒ‡ä»¤ï¼ˆ*.bz)ï¼Œx-é‡Šæ”¾ï¼Œv-é‡Šæ”¾æ—¶çš„ä¿¡æ¯ï¼Œf-æ–‡ä»¶
cd samtools-1.17

# é…ç½®ä¿¡æ¯
./configure --prefix=$(pwd)#æºç çš„å®‰è£…ä¸€èˆ¬ç”±3ä¸ªæ­¥éª¤ç»„æˆï¼šé…ç½®ï¼ˆconfigureï¼‰ã€ç¼–è¯‘ï¼ˆmakeï¼‰ã€å®‰è£…ï¼ˆmake installï¼‰ï¼ŒConfigureæ˜¯ä¸€ä¸ªå¯æ‰§è¡Œè„šæœ¬ï¼Œå®ƒæœ‰å¾ˆå¤šé€‰é¡¹ï¼Œä½¿ç”¨å‘½ä»¤./configure â€“helpè¾“å‡ºã€‚ configure --prefix=/â€çš„ä½œç”¨æ˜¯ï¼šç¼–è¯‘çš„æ—¶å€™ç”¨æ¥æŒ‡å®šç¨‹åºå­˜æ”¾è·¯å¾„

# å¼€å§‹ç¼–è¯‘
make -j 4

# å¯¼å…¥ä¸´æ—¶ç¯å¢ƒå˜é‡
export PATH="$(pwd):$PATH"

#æ£€æŸ¥samtoolsæ˜¯å¦å®‰è£…æˆåŠŸ
samtools

#å‡ºç°ä»¥ä¸‹ç•Œé¢
Program: samtools (Tools for alignments in the SAM format)
Version: 1.17 (using htslib 1.17)

Usage:   samtools <command> [options]

Commands:
  -- Indexing
     dict           create a sequence dictionary file
     faidx          index/extract FASTA
     fqidx          index/extract FASTQ
     index          index alignment
```

## 2.8 HTseq
å¯¹æ¯”å¯¹åçš„æ–‡ä»¶è¿›è¡Œreadè®¡æ•°
```
pip install -i https://pypi.tuna.tsinghua.edu.cn/simple HTseq #æœ‰é—®é¢˜
```
## 2.9 R
[å®˜ç½‘](https://www.r-project.org)

- ä»‹ç»
  
Rè¯­è¨€ä¸­é›†åˆäº†å¤šç§ç”Ÿç‰©ä¿¡æ¯å­¦çš„åˆ†æå·¥å…·ï¼Œå…¶ä¸­RNA-seqåˆ†æçš„å·¥å…·æ›´æ˜¯ä¸°å¯Œï¼ŒRè¯­è¨€æœ€æ“…é•¿ç»Ÿè®¡å­¦åˆ†æï¼Œè¿™é‡Œåœ¨åç»­çš„åŸºå› è¡¨è¾¾é‡åˆ†æï¼Œå·®å¼‚åˆ†æä»¥åŠä½œå›¾æ—¶å€™ä¼šç”¨åˆ°
- ä¸‹è½½

ç‚¹å‡»å·¦ä¸Šè§’```CRAN```ï¼Œå¾€ä¸‹æ‹‰æ‰¾åˆ°```china```çš„ç«™ç‚¹ï¼Œå°±é€‰ç¬¬ä¸€ä¸ªæ¸…åçš„ç«™ç‚¹ï¼Œç‚¹å‡»è¿›å»ï¼Œä¹‹åæœ€ä¸Šé¢æœ‰ä¸‰ä¸ªå¯¹åº”ç³»ç»Ÿçš„å®‰è£…åŒ…ï¼ŒæŒ‰ç…§è‡ªå·±çš„ç³»ç»Ÿä¸‹è½½ï¼Œè¿™é‡Œç‚¹å‡»```Windows```ï¼Œç‚¹å‡»```install R for the first time```ååœ¨æ–°ç•Œé¢ç‚¹å‡»```Download R-4.2.2 for Windows```å°±å¼€å§‹ä¸‹è½½äº†ï¼Œä¸‹è½½ä¹‹å```åŒå‡»```å®‰è£…åŒ…å®‰è£…ã€‚
- brew
```
brew install r
```
## 2.10 Rstudio
å› ä¸ºRè‡ªèº«å¸¦çš„ç•Œé¢ä½¿ç”¨èµ·æ¥ä¸æ˜¯ç‰¹åˆ«æ–¹ä¾¿å’Œç¾è§‚ï¼Œè¿™é‡Œä½¿ç”¨Rstudioæ¥å¯¹Rçš„æ˜¾ç¤ºæ•ˆæœè¿›è¡Œæå‡ï¼Œé™¤æ­¤ä¹‹å¤–è¿˜æœ‰å…¶ä»–åŠŸèƒ½ã€‚

è¿›å…¥ç½‘ç«™ï¼šhttps://www.rstudio.com/products/rstudio/download/

ç‚¹å‡»```RStudio Desktop Open Source License``` ä¸‹é¢çš„DOWNLOADï¼Œä¹‹åé€‰æ‹©å¯¹åº”çš„ç‰ˆæœ¬çš„ï¼Œè¿™é‡Œé€‰æ‹©```Windows 10/11 OS```ç‰ˆæœ¬
```RSTUDIO-2023.03.1-446.EXE```ä¸‹è½½å®Œæˆä¹‹ååŒå‡»å®‰è£…ã€‚

## 2.11 parallel
parallelæ˜¯è¿›è¡Œå¤šçº¿ç¨‹è¿è¡Œçš„å·¥å…·ï¼Œå¹¶è¡Œè¿è¡Œå¯ä»¥æå‡æ•ˆç‡ï¼ŒèŠ‚çœæ—¶é—´
```
brew install parallel
```

## Ballgown[å¯é€‰]
æ˜¯Rè¯­è¨€ä¸­åŸºå› å·®å¼‚è¡¨è¾¾åˆ†æçš„å·¥å…·ï¼Œèƒ½åˆ©ç”¨RNA-Seqå®éªŒçš„æ•°æ®(StringTie, RSEM, Cufflinks)çš„ç»“æœé¢„æµ‹åŸºå› ã€è½¬å½•æœ¬çš„å·®å¼‚è¡¨è¾¾ã€‚


# 3. æ•°æ®ä¸‹è½½

è¿™é‡Œä½¿ç”¨å¤§é¼ çš„æµ‹åºæ•°æ®ä½œä¸ºæµ‹è¯•ã€‚å¤§é¼ åˆå«å¤§ç™½é¼ ï¼ˆRatï¼ŒRattus norvegicusï¼‰ï¼Œåœ¨ENSEMBLå’ŒUCSCä¸­å‡æœ‰å…¶åŸºå› ç»„æ•°æ®ã€‚

## 3.1 å‚è€ƒæ•°æ®
ä¸‹è½½åŸºå› ç»„åºåˆ—

[Ensembl](https://asia.ensembl.org/index.html)

åœ¨å·¦ä¾§```All genomes```ä¸­ï¼Œé€‰æ‹©ç‰©ç§```Rat```; åœ¨å·¦ä¾§```Download DNA sequence (FASTA)``` ä¸‹è½½åŸºå› ç»„åºåˆ—æ•°æ®; åœ¨å³ä¾§çš„```Download GTF or GFF3 (files for genes, cDNAs, ncRNA, proteins)```ä¸‹è½½åŸºå› æ³¨é‡Šæ–‡ä»¶

![image](https://github.com/QRunMiao/RNA-seq-study/blob/picture/1676251978571.jpg)

- [ensmbleçš„readme](https://ftp.ensembl.org/pub/release-109/fasta/rattus_norvegicus/dna/README)

## ensmble FILE NAMES
é¦–å…ˆè§„å®šäº†æ–‡ä»¶å‘½åè§„å¾‹ï¼š<species>.<assembly>.<sequence type>.<id type>.<id>.fa.gz

ç¬¬ä¸€éƒ¨åˆ†<species>æ˜¯ç‰©ç§åï¼›

ç¬¬äºŒéƒ¨åˆ†<assembly>æ˜¯åŸºå› ç»„ç‰ˆæœ¬å·ï¼›

ç¬¬ä¸‰éƒ¨åˆ†<sequence type>æ˜¯åºåˆ—ç±»å‹ï¼š

Â· dnaï¼šåŸåŸæœ¬æœ¬çš„DNAåºåˆ—

Â· dna_rmï¼ˆhard_maskï¼‰ï¼šåˆ©ç”¨RepeatMaskerå·¥å…·å°†é‡å¤åŒºåŸŸå’Œä½å¤æ‚åº¦åŒºåŸŸæ ‡è®°æˆä¸€ä¸²N

Â· dna_smï¼ˆsoft_maskï¼‰ï¼šæ‰€æœ‰é‡å¤åŒºåŸŸå’Œä½å¤æ‚åº¦åŒºåŸŸæ›¿æ¢ä¸ºå°å†™çš„ç¢±åŸº

ç¬¬å››éƒ¨åˆ†<id type>æ˜¯åŸºå› ç»„ç±»å‹ï¼š

Â· chromosomeï¼šæŸ“è‰²ä½“åºåˆ—

Â· nonchromosomalï¼šå­˜å‚¨ç›®å‰æ²¡æœ‰æ¯”å¯¹åˆ°æŸ“è‰²ä½“çš„DNAåºåˆ—

Â· seqlevelï¼šä¸€èˆ¬æ˜¯scaffolds, chunks æˆ–clonesï¼Œè¿˜æœªç»„è£…åˆ°æŸ“è‰²ä½“æ ‡å‡†çš„åºåˆ—ä¿¡æ¯


ç¬¬äº”éƒ¨åˆ†<id>æ˜¯å®é™…çš„åºåˆ—ç¼–å·ï¼Œæ¯”å¦‚å±äºå“ªæ¡æŸ“è‰²ä½“æˆ–è€…å±äºå“ªä¸ªscaffold, chunk æˆ–clone

æœ€åè¿™ä¸ªfaæ–‡ä»¶ä»¥.gzå½¢å¼å‹ç¼©ä¿å­˜

Â· å¦‚äººç±»1å·æŸ“è‰²ä½“çš„åŸºå› ç»„åºåˆ—ï¼š```Homo_sapiens.GRCh37.dna.chromosome.1.fa.gz```

- TOPLEVELï¼š è¿™æ ·çš„æ•°æ®åŒ…å«äº†æ‰€æœ‰çš„åºåˆ—åŒºåŸŸï¼ˆæ¯”å¦‚æŸ“è‰²ä½“ã€éæŸ“è‰²ä½“ä»¥åŠç”¨å¤§é‡Nå¡«å……çš„å•å€å‹haplotypes-(æŒ‡ä¸€ä¸ªæŸ“è‰²å•ä½“é‡Œé¢å…·æœ‰ç»Ÿè®¡å­¦å…³è”æ€§çš„ä¸€ç±»å•æ ¸è‹·é…¸å¤šæ€æ€§ï¼ˆSNPsï¼‰æˆ–åŸºå› ç»„è¡¥ä¸patchesåŒºåŸŸï¼‰ï¼Œæ¯”å¦‚ï¼š```Homo_sapiens.GRCh37.dna.toplevel.fa.gz```

- PRIMARY ASSEMBLYï¼š åœ¨ä¸Šé¢toplevelçš„åŸºç¡€ä¸Šï¼Œæ’é™¤äº†å•å€å‹æˆ–åŸºå› ç»„è¡¥ä¸åŒºåŸŸã€‚å¦‚æœçœ‹åˆ°ç›®å½•ä¸­ä¸å­˜åœ¨è¿™ç§ç±»å‹çš„æ•°æ®ï¼ˆæ¯”å¦‚è¿™é‡Œæœè‡å°±æ²¡æœ‰ï¼Œè€Œäººç±»çš„åŸºå› ç»„æ•°æ®å°±å­˜åœ¨ï¼‰ï¼Œé‚£ä¹ˆå°±æ„å‘³ç€åŸºå› ç»„ä¸åŒ…å«å•å€å‹æˆ–åŸºå› ç»„è¡¥ä¸åŒºåŸŸï¼Œå…¶å®ä¹Ÿå°±æ˜¯ç­‰åŒäºTOPLEVEL

- special æ³¨æ„
ä¸€äº›æŸ“è‰²ä½“æ˜¯å•å€ä½“ï¼Œä¾‹å¦‚äººç±»çš„Xå’ŒYæŸ“è‰²ä½“
ä¸ºäº†æ¯”å¯¹æ—¶èƒ½æ­£ç¡®è¾“å‡ºæŠ¥å‘Šï¼Œè¿™äº›å•å€ä½“çš„assemblyå’ŒpatchåŒºåŸŸéƒ½ä¼šè¡¥ä¸ŠåŒç­‰æ•°é‡çš„N


## ä¸‹è½½

ä¸€èˆ¬ä¸‹è½½toplevel.fa.gzæ–‡ä»¶ï¼Œä¸ºå‚è€ƒåŸºå› ç»„å®Œæ•´æ–‡ä»¶

åŸºå› ç»„æ•°æ®è¯´æ˜

æ ¼å¼ä¸º.fastaï¼Œæ ¼å¼ä¸ºï¼š
```
>seq_id
AGCTGAGCTAGCTACGGAGCTGAC
ACGACTGATCTGACGTTGATCGTT
```

æ–‡ä»¶ä¸­ä»¥>å¼€å¤´çš„æ˜¯åºåˆ—çš„åç§°ï¼Œä¸‹é¢æ¥ç€çš„ATGCæ˜¯è¿™æ¡åºåˆ—ä¿¡æ¯ï¼ŒåŸºå› ç»„fastaæ–‡ä»¶è®°å½•å¤§é¼ çš„æ‰€æœ‰çš„è¢«æµ‹å¾—çš„æŸ“è‰²ä½“çš„åºåˆ—ä¿¡æ¯ï¼Œç›®å‰å¤§é¼ çš„åŸºå› ç»„æµ‹åºç‰ˆæœ¬åˆ°äº†7.2

å¯ä»¥ç›´æ¥åœ¨ç½‘é¡µä¸‹è½½ï¼Œä¹Ÿå¯ç”¨ä»£ç 

```
cd /mnt/c/shengxin/project/rat/genome

wget https://ftp.ensembl.org/pub/release-109/fasta/rattus_norvegicus/dna/Rattus_norvegicus.mRatBN7.2.dna.toplevel.fa.gz

gzip -d Rattus_norvegicus.mRatBN7.2.dna.toplevel.fa.gz
```
## æ”¹å
ï¼ˆæ–¹ä¾¿åé¢ä½¿ç”¨ï¼Œåå­—å¤ªé•¿ä¸€æ¥ä¸æ–¹ä¾¿è¾“å…¥ï¼ŒäºŒæ¥å¯èƒ½ä¼šè¾“é”™ï¼‰

```mv Rattus_norvegicus.mRatBN7.2.dna.toplevel.fa R7.2.fa```

- cat & grep ä¸€èµ·ä½¿ç”¨å¯ä»¥å®ç°åœ¨æ–‡æœ¬å†…æŸ¥æ‰¾å…³é”®è¯

```cat æ–‡ä»¶ | grep å…³é”®å­—``` 

ä¸‹è½½å¾—åˆ°çš„åŸºå› ç»„æ–‡ä»¶å¯ä»¥æŸ¥çœ‹ä¸€ä¸‹åŒ…å«å“ªäº›æŸ“è‰²ä½“ï¼Œç¡®è®¤æ–‡ä»¶æ˜¯å¦ä¸‹è½½æ­£ç¡®

```cat R7.2.fa | grep "^>" ```

ç»“æœï¼šé™¤äº†```1-20å·+X+Y+MT```ä¹‹å¤–è¿˜æœ‰å¾ˆå¤šåˆ«çš„IDåã€‚è¿™äº›éƒ½æ˜¯```scaffold```

```
>1 dna:primary_assembly primary_assembly:mRatBN7.2:1:1:260522016:1 REF
>2 dna:primary_assembly primary_assembly:mRatBN7.2:2:1:249053267:1 REF
>3 dna:primary_assembly primary_assembly:mRatBN7.2:3:1:169034231:1 REF
>4 dna:primary_assembly primary_assembly:mRatBN7.2:4:1:182687754:1 REF
>5 dna:primary_assembly primary_assembly:mRatBN7.2:5:1:166875058:1 REF
>6 dna:primary_assembly primary_assembly:mRatBN7.2:6:1:140994061:1 REF
>7 dna:primary_assembly primary_assembly:mRatBN7.2:7:1:135012528:1 REF
>8 dna:primary_assembly primary_assembly:mRatBN7.2:8:1:123900184:1 REF
>9 dna:primary_assembly primary_assembly:mRatBN7.2:9:1:114175309:1 REF
>10 dna:primary_assembly primary_assembly:mRatBN7.2:10:1:107211142:1 REF
>11 dna:primary_assembly primary_assembly:mRatBN7.2:11:1:86241447:1 REF
>12 dna:primary_assembly primary_assembly:mRatBN7.2:12:1:46669029:1 REF
>13 dna:primary_assembly primary_assembly:mRatBN7.2:13:1:106807694:1 REF
>14 dna:primary_assembly primary_assembly:mRatBN7.2:14:1:104886043:1 REF
>15 dna:primary_assembly primary_assembly:mRatBN7.2:15:1:101769107:1 REF
>16 dna:primary_assembly primary_assembly:mRatBN7.2:16:1:84729064:1 REF
>17 dna:primary_assembly primary_assembly:mRatBN7.2:17:1:86533673:1 REF
>18 dna:primary_assembly primary_assembly:mRatBN7.2:18:1:83828827:1 REF
>19 dna:primary_assembly primary_assembly:mRatBN7.2:19:1:57337602:1 REF
>20 dna:primary_assembly primary_assembly:mRatBN7.2:20:1:54435887:1 REF
>X dna:primary_assembly primary_assembly:mRatBN7.2:X:1:152453651:1 REF
>Y dna:primary_assembly primary_assembly:mRatBN7.2:Y:1:18315841:1 REF
>MT dna:primary_assembly primary_assembly:mRatBN7.2:MT:1:16313:1 REF
>MU150191.1 dna:primary_assembly primary_assembly:mRatBN7.2:MU150191.1:1:1794995:1 REF
```

```
# é¦–å…ˆå°†ä¹‹å‰çš„åç§°æ›´æ”¹ä¸€ä¸‹
mv R7.2.fa R7.2.raw.fa
mvï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šmove fileï¼‰å‘½ä»¤ç”¨æ¥ä¸ºæ–‡ä»¶æˆ–ç›®å½•æ”¹åã€æˆ–å°†æ–‡ä»¶æˆ–ç›®å½•ç§»å…¥å…¶å®ƒä½ç½®

# ç„¶åå»é™¤æŸ“è‰²ä½“ç¼–å·åçš„æè¿°ä¿¡æ¯
$ cat R7.2.raw.fa | perl -n -e 'if(m/^>(.+?)(?:\s|$)/){ print ">$1\n";}else{print}' > R7.2.fa
#ä»¥ ">" å¼€å¤´ï¼Œè¡¨ç¤ºæ˜¯ä¸€ä¸ªåºåˆ—æ ‡é¢˜è¡Œã€‚ä½¿ç”¨æ­£åˆ™è¡¨è¾¾å¼ m/^>(.+?)(?:\s|$)/ åŒ¹é…æ ‡é¢˜ï¼Œå¹¶æå–å‡ºåŒ¹é…ç»“æœä¸­çš„ç¬¬ä¸€ä¸ªæ•è·ç»„ï¼ˆå³æ ‡é¢˜çš„éƒ¨åˆ†ï¼‰ï¼›må¤šè¡ŒåŒ¹é…ï¼Œå¦‚æœåŒ¹é…åˆ°äº†ä»¥>å¼€å¤´å¤šä¸ªå­—æ¯ç©ºæ ¼æˆ–è€…$ï¼Œ"$1" æ˜¯åŒ¹é…åˆ°çš„æ ‡é¢˜éƒ¨åˆ†ï¼Œæ·»åŠ  ">" åå¹¶æ¢è¡Œè¾“å‡ºã€‚å¦åˆ™ï¼Œå³ä¸ºåºåˆ—å†…å®¹è¡Œï¼Œç›´æ¥ä½¿ç”¨ print å‘½ä»¤å°†å…¶è¾“å‡ºã€‚

# åˆ é™¤
$ rm R7.2.raw.fa
```

å¯ä»¥ä½¿ç”¨è„šæœ¬ç»Ÿè®¡æ¯ä¸€æ¡æŸ“è‰²ä½“çš„é•¿åº¦
```
$ cat R7.2.fa | perl -n -e '#-n å‚æ•°å‘Šè¯‰ perl åœ¨æ¯ä¸€è¡Œä¸Šæ‰§è¡Œç»™å®šçš„ä»£ç ã€‚-e å‚æ•°æŒ‡å®šäº†è¦æ‰§è¡Œçš„ä»£ç å—
     #é™¤å»æœ«å°¾çš„å›è½¦ç¬¦ã€æ¢è¡Œç¬¦
     s/\r?\n//;
     # å¾—åˆ°åºåˆ—çš„åç§°,
    if(m/^>(.+?)\s*$/){
       $title = $1;
        push @t, $title;#å¦‚æœå½“å‰è¡Œä»¥ ">" å¼€å¤´ï¼Œè¡¨ç¤ºæ˜¯ä¸€ä¸ªåºåˆ—æ ‡é¢˜è¡Œï¼Œä½ æå–å‡ºæ ‡é¢˜ï¼Œå¹¶å°†å®ƒæ·»åŠ åˆ° @t æ•°ç»„ä¸­ã€‚
    }elsif(defined $title){
     # å¦åˆ™ï¼Œå¦‚æœ $title å·²ç»å®šä¹‰ï¼ˆå³å‰é¢å·²ç»é‡åˆ°äº†åºåˆ—æ ‡é¢˜ï¼‰ï¼Œåˆ™è®¡ç®—å½“å‰è¡Œçš„é•¿åº¦ï¼Œå¹¶å°†å…¶åŠ åˆ°ç›¸åº”çš„æ ‡é¢˜é•¿åº¦ä¸­ ($title_len)ã€‚
        $title_len{$title} += length($_); 
    }
     # æœ€åæ‰“å°å‡ºä¿¡æ¯æ¥
    END{
        for my $title (@t){
            print "$title","\t","$title_len{$title}","\n";
        }åœ¨ END å—ä¸­ï¼Œéå† @t æ•°ç»„ä¸­çš„æ ‡é¢˜ï¼Œå¹¶æ‰“å°å‡ºæ¯ä¸ªæ ‡é¢˜å’Œå…¶å¯¹åº”çš„é•¿åº¦ã€‚
    }
'

```

ç»“æœ
```
1       260522016
2       249053267
3       169034231
4       182687754
5       166875058
6       140994061
7       135012528
8       123900184
9       114175309
10      107211142
11      86241447
12      46669029
13      106807694
14      104886043
15      101769107
16      84729064
17      86533673
18      83828827
19      57337602
20      54435887
X       152453651
Y       18315841
MT      16313
MU150191.1      1794995
MU150189.1      1402623
MU150194.1      648519
MU150190.1      573231
MU150195.1      529129
JACYVU010000493.1       444596
MU150193.1      383091
```

è¿™é‡Œä¸ºäº†æ–¹ä¾¿æ¼”ç¤ºï¼Œç›´æ¥ç”¨1å·æŸ“è‰²ä½“çš„åŸºå› ç»„ä½œä¸ºåç»­æ¯”å¯¹ã€‚
```
cat R7.2.fa | perl -n -e '
  if(m/^>/){
    if(m/>1$/){
      $title = 1; # title=1ï¼Œå³ä¸€å·æŸ“è‰²ä½“
    }else{
      $title = 0;
    }
  }else{
    push @s, $_ if $title;
  }
  END{
    printf ">1\n%s", join("", @s);
  }
' > R7.2.chr1.fa
```

- ä¸‹è½½åŸºå› ç»„ç´¢å¼•æ–‡ä»¶(å¯é€‰)
  
åŸºå› ç»„ç´¢å¼•æ–‡ä»¶å°±åƒä¸€æœ¬ä¹¦çš„ç›®å½•ï¼Œé‡Œé¢é›†æˆäº†æ¯”å¯¹è½¯ä»¶çš„ç®—æ³•ï¼Œæ‰€ä»¥è™½ç„¶åŸºå› ç»„åºåˆ—æ˜¯ä¸€æ ·çš„ï¼Œä½†ä¸åŒçš„æ¯”å¯¹è½¯ä»¶éœ€è¦ä¸åŒçš„åŸºå› ç»„ç´¢å¼•æ–‡ä»¶ã€‚

åœ¨[hisat2](http://daehwankimlab.github.io/hisat2/) å®˜ç½‘ä¸Šå¯ä»¥æ‰¾åˆ°ç°æˆçš„å·²ç»å»ºç«‹å¥½ç´¢å¼•çš„å¤§é¼ åŸºå› ç»„æ–‡ä»¶ ï¼ˆåœ¨```download```é‡Œé¢æ‰¾ï¼‰ï¼Œå¦‚æœç”µè„‘é…ç½®ä¸€èˆ¬å»ºè®®ç›´æ¥ä¸‹è½½å¥½ç´¢å¼•æ–‡ä»¶ï¼Œå¯ä»¥ç›´æ¥ä¸‹è½½è¿™ä¸ªç´¢å¼•æ–‡ä»¶ï¼ˆå› ä¸ºå»ºç«‹ç´¢å¼•æ–‡ä»¶æ—¶é—´è¾ƒé•¿1ä¸ªå°æ—¶ä»¥ä¸Šï¼‰ï¼Œè¿™ä¸ªç´¢å¼•æ–‡ä»¶æ˜¯å¯ä»¥è‡ªå·±ç”¨å‘½ä»¤åŸºäºä¹‹å‰ä¸‹è½½çš„åŸºå› ç»„æ–‡ä»¶è‡ªè¡Œå»ºç«‹çš„ã€‚

æ–¹æ³•1ï¼š

åœ¨```hisat2```å®˜ç½‘ä¸Šå¯ä»¥æ‰¾åˆ°ç°æˆçš„å·²ç»å»ºç«‹å¥½ç´¢å¼•çš„å¤§é¼ åŸºå› ç»„æ–‡ä»¶, ç‚¹å‡» https://genome-idx.s3.amazonaws.com/hisat/rn6_genome.tar.gz

æ–¹æ³•2ï¼šé€šè¿‡ä»£ç ä¸‹è½½

```
cd /mnt/c/shengxin/project/rat/genome
wget https://genome-idx.s3.amazonaws.com/hisat/rn6_genome.tar.gz
gzip -d rn6_genome.tar.gz
```

- ä¸‹è½½æ³¨é‡Šä¿¡æ¯(gff3 æ ¼å¼)

GFF (general feature format): å¯ä»¥ç”¨äºä»»ä½•åŸºå› ç»„æ³¨é‡Šçš„å­˜å‚¨

GTF (gene transfer format): ä¸¥æ ¼çš„ç”¨äºåŸºå› æ³¨é‡Šä¿¡æ¯çš„å­˜å‚¨

[gff/gtfå„åˆ—ä»£è¡¨å«ä¹‰](https://www.jianshu.com/p/c284a6b4e1c6)
```
cd /mnt/c/shengxin/project/rat/annotation
wget https://ftp.ensembl.org/pub/release-109/gff3/rattus_norvegicus/Rattus_norvegicus.mRatBN7.2.109.gff3.gz
gzip -d Rattus_norvegicus.mRatBN7.2.109.gff3.gz

# åŒæ ·çš„ä¹Ÿæ”¹å
mv Rattus_norvegicus.mRatBN7.2.109.gff3 R7.2.gff

# ä½¿ç”¨headæŸ¥çœ‹éƒ¨åˆ†
head R7.2.gff
```

R7.2.gffæŸ¥çœ‹ç»“æœ
```
##gff-version 3
##sequence-region   1 1 260522016
##sequence-region   10 1 107211142
##sequence-region   11 1 86241447
##sequence-region   12 1 46669029
##sequence-region   13 1 106807694
##sequence-region   14 1 104886043
##sequence-region   15 1 101769107
##sequence-region   16 1 84729064
##sequence-region   17 1 86533673
```

- ä¸‹è½½ gtf æ ¼å¼
```  
cd /mnt/c/shengxin/project/rat/annotation
wget https://ftp.ensembl.org/pub/release-109/gtf/rattus_norvegicus/Rattus_norvegicus.mRatBN7.2.109.gtf.gz
gzip -d Rattus_norvegicus.mRatBN7.2.109.gtf.gz

# åŒæ ·çš„ä¹Ÿæ”¹å
mv Rattus_norvegicus.mRatBN7.2.109.gtf R7.2.gtf

# ä½¿ç”¨headæŸ¥çœ‹éƒ¨åˆ†
head R7.2.gtf
```

- æ³¨é‡Šæ•°æ®è¯´æ˜

æ³¨é‡Šgffæ–‡ä»¶çš„æ ·ä¾‹ï¼š
```
#!genome-build mRatBN7.2
#!genome-version mRatBN7.2
#!genome-date 2020-11
#!genome-build-accession GCA_015227675.2
#!genebuild-last-updated 2021-02
1       ensembl gene    36112690        36122387        .       -       .       gene_id "ENSRNOG00000066169"; gene_version "1"; gene_source "ensembl"; gene_biotype "protein_coding";
1       ensembl transcript      36112690        36122387        .       -       .       gene_id "ENSRNOG00000066169"; gene_version "1"; transcript_id "ENSRNOT00000101581"; transcript_version "1"; gene_source "ensembl"; gene_biotype "protein_coding"; transcript_source "ensembl"; transcript_biotype "protein_coding"; tag "Ensembl_canonical";
1       ensembl exon    36122324        36122387        .       -       .       gene_id "ENSRNOG00000066169"; gene_version "1"; transcript_id "ENSRNOT00000101581"; transcript_version "1"; exon_number "1"; gene_source "ensembl"; gene_biotype "protein_coding"; transcript_source "ensembl"; transcript_biotype "protein_coding"; exon_id "ENSRNOE00000618632"; exon_version "1"; tag "Ensembl_canonical";
1       ensembl CDS     36122324        36122387        .       -       0       gene_id "ENSRNOG00000066169"; gene_version "1"; transcript_id "ENSRNOT00000101581"; transcript_version "1"; exon_number "1"; gene_source "ensembl"; gene_biotype "protein_coding"; transcript_source "ensembl"; transcript_biotype "protein_coding"; protein_id "ENSRNOP00000083062"; protein_version "1"; tag "Ensembl_canonical";
1       ensembl exon    36121478        36121512        .       -       .       gene_id "ENSRNOG00000066169"; gene_version "1"; transcript_id "ENSRNOT00000101581"; transcript_version "1"; exon_number "2"; gene_source "ensembl"; gene_biotype "protein_coding"; transcript_source "ensembl"; transcript_biotype "protein_coding"; exon_id "ENSRNOE00000610554"; exon_version "1"; tag "Ensembl_canonical";
```
gtfæ–‡ä»¶å¼€å¤´æè¿°äº†è¿™ä¸ªæ³¨é‡Šæ•°æ®çš„åŸºæœ¬ä¿¡æ¯ï¼Œæ¯”å¦‚ç‰ˆæœ¬å·ï¼Œæ›´æ–°æ—¶é—´ï¼Œç»„è£…çš„NCBIçš„Assemblyç¼–å·ç­‰ç­‰ï¼Œåé¢æ¯ä¸€è¡Œè¡¨ç¤ºæè¿°ä¿¡æ¯ï¼Œè¯´æ˜äº†åœ¨å“ªæ¡æŸ“è‰²ä½“çš„ä»€ä¹ˆä½ç½®æ˜¯ä»€ä¹ˆä¸œè¥¿ã€‚æ¯”å¦‚ç¬¬6è¡Œçš„è¡¨ç¤ºåœ¨1å·æŸ“è‰²ä½“è´Ÿé“¾36112690-36122387èŒƒå›´ä¸Šæœ‰ä¸€ä¸ªåŸºå› ç¼–å·ä¸ºENSRNOG00000066169çš„åŸºå› 

## 3.2 æµ‹è¯•æ•°æ®ï¼ˆå®éªŒæ•°æ®ï¼‰
ä»NCBIä¸ŠæŸ¥æ‰¾ç›¸å…³çš„RNA-seqæ•°æ®è¿›è¡Œä¸‹è½½ï¼Œåœ¨GEOæ•°æ®åº“ä¸­æ‰¾äº†ä¸€ä¸ªæ•°æ®```GSE72960```ï¼Œå¯¹åº”çš„SRPæ•°æ®ä¸º```SRP063345```

SRAå’ŒGEO (Gene Expression Omnibus database) éƒ½æ˜¯NCBIé‡Œé¢çš„ï¼ŒSRAåŸå§‹æ•°æ®GEOç»è¿‡ä¸€å®šå¤„ç†çš„æ•°æ®

GEO-æ”¶å½•äº†ä¸–ç•Œå„å›½ç ”ç©¶æœºæ„æäº¤çš„é«˜é€šé‡åŸºå› è¡¨è¾¾æ•°æ®ï¼Œä¹Ÿå°±æ˜¯è¯´åªè¦æ˜¯ç›®å‰å·²ç»å‘è¡¨çš„è®ºæ–‡ï¼Œè®ºæ–‡ä¸­æ¶‰åŠåˆ°çš„åŸºå› è¡¨è¾¾æ£€æµ‹çš„æ•°æ®éƒ½å¯ä»¥é€šè¿‡è¿™ä¸ªæ•°æ®åº“ä¸­æ‰¾åˆ°ã€‚

é€šè¿‡[NCBI - GEO](https://www.ncbi.nlm.nih.gov/geo/)è¿›å…¥ï¼Œæœç´¢æ¡†ä¸­è¾“å…¥```GSE72960```ï¼ˆä¸€èˆ¬åœ¨æ–‡çŒ®ä¸­å¯æŸ¥åˆ°ï¼‰

ä¹‹ååœ¨ä¸‹æ–¹å‡ºç°äº†è¿™ä¸ªåŸºå› è¡¨è¾¾æ•°æ®é›†çš„æè¿°ä¿¡æ¯ï¼Œæ¯”å¦‚æ ·æœ¬æäº¤æ—¥æœŸï¼Œæ ·æœ¬æ–‡çŒ®æ¥æºï¼Œæ•°æ®æäº¤äººçš„ä¿¡æ¯ï¼Œæ ·æœ¬æµ‹åºæ ·æœ¬æ•°é‡ï¼Œå¯¹åº”çš„ç¼–å·ç­‰ç­‰

æˆ‘ä»¬ç›´æ¥ç‚¹å‡»æœ€ä¸‹é¢çš„SRA Run selectorè¿™ä¸ªé‡Œé¢åŒ…å«äº†è¿™8ä¸ªæµ‹åºæ ·æœ¬çš„æµ‹åºä¿¡æ¯ä»¥åŠæ–‡ä»¶SRAç¼–å·ï¼Œé€šè¿‡è¿™ä¸ªç¼–å·å°±å¯ä»¥ä¸‹è½½æµ‹åºæ•°æ®ã€‚

å°†åˆšæ‰åœ¨Run selectorä¸­æŸ¥æ‰¾åˆ°çš„æ•°æ®çš„ç¼–å·å¤åˆ¶ä¸‹æ¥ï¼Œä¹‹åä¸‹è½½æµ‹åºæ•°æ®ï¼Œä¸‹è½½è„šæœ¬å¦‚ä¸‹ï¼Œè¿™é‡Œæ˜¯é‡‡ç”¨SRAtoolkitå·¥å…·åŒ…ä¸­çš„prefetchå·¥å…·ï¼Œå¦‚æœéƒ¨åˆ†æ•°æ®ä¸‹è½½å¤±è´¥ï¼Œé‚£ä¹ˆå†æ¬¡æ‰§è¡Œä¸‹é¢çš„ä»£ç 
```
cd /mnt/c/shengxin/project/rat/sra
# åå°ä¸‹è½½
$ nohup prefetch SRR2190795 SRR224018{2..7} SRR2240228 -o . &
# nohup è‹±æ–‡å…¨ç§° no hang upï¼ˆä¸æŒ‚èµ·ï¼‰ï¼Œç”¨äºåœ¨ç³»ç»Ÿåå°ä¸æŒ‚æ–­åœ°è¿è¡Œå‘½ä»¤ï¼Œé€€å‡ºç»ˆç«¯ä¸ä¼šå½±å“ç¨‹åºçš„è¿è¡Œã€‚
# -o . & ä¸‹è½½å®Œè¾“å‡ºåˆ°å½“å‰ç›®å½•ï¼Œ& ä¸€èˆ¬æ²¡æœ‰è¿è¡Œå®Œã€è¾“å…¥ç©ºæ ¼ä¼šå¡ä½ï¼Œåªæœ‰è¿è¡Œå®Œæ‰ä¼šå†å‡ºç°ä¸»æœºåç§°ï¼›ä½†æ˜¯åŠ ä¸Š&åå¯ä»¥æ”¾åœ¨åå°ä¸‹è½½è¿è¡Œ
#prefetch æ‰¹é‡ä¸‹è½½æ•°æ®

```
ä¸‹è½½å®Œæˆä¹‹åå¹¶ä¸æ˜¯ä¹‹å‰è¯´çš„.fastq.gzæ ¼å¼çš„æ–‡ä»¶ï¼Œè€Œæ˜¯.sraæ–‡ä»¶ï¼Œè¿™é‡Œè¿›è¡Œæ ¼å¼è½¬æ¢ï¼Œè¿™é‡Œè¿˜æ˜¯ä½¿ç”¨SRAtoolkitå·¥å…·åŒ…ï¼Œä½†æ˜¯æ˜¯é‡Œé¢çš„fastq-dumpå·¥å…·ï¼Œä½¿ç”¨å®ƒæ¥è¿›è¡Œæ ¼å¼è½¬åŒ–
```
# å°†sraæ–‡ä»¶è½¬åŒ–ä¸ºfastqæ–‡ä»¶ä¹‹åå‹ç¼©ä¸ºgzæ–‡ä»¶

# --gzip æŠŠç”Ÿæˆçš„fastqæ–‡ä»¶å‹ç¼©ä¸ºgzæ ¼å¼ï¼ŒèŠ‚çœå†…å­˜
parallel -j 4 "
    fastq-dump --split-3 --gzip {1}#fastq-dumpä»SRAæ–‡ä»¶ä¸­æå–fastqæ–‡ä»¶.{1} è¡¨ç¤ºå ä½ç¬¦ï¼Œä¼šè¢«åé¢çš„æ–‡ä»¶åæ›¿ä»£ã€‚
" ::: $(ls *.sra)#::: åé¢æ˜¯è¦éå†çš„å‚æ•°é›†åˆ

# åˆ é™¤sraæ–‡ä»¶
rm *.sra
```

fastqæ ¼å¼ä»‹ç»
```
  cd /mnt/c/shengxin/project/rat/sequence
  gzip -d -c SRR2190795.fastq.gz | head -n 20
@SRR2190795.1 HWI-ST1147:240:C5NY7ACXX:1:1101:1320:2244 length=100
ATGCTGGGGGCATTAGCATTGGGTACTGAATTATTTTCAGTAAGAGGGAAAGAATCCATCTCCNNNNNNNNNNNNNNNNNNNNNNAAANAAAAATAAAAT
+SRR2190795.1 HWI-ST1147:240:C5NY7ACXX:1:1101:1320:2244 length=100
CCCFFFFFHHHHHJIJJJJJJJJDHHJJJIJJJJJIJJJJJJJJJJJJJJJJJJJJJJJJJHH#####################################
@SRR2190795.2 HWI-ST1147:240:C5NY7ACXX:1:1101:1598:2247 length=100
AACTTCGGTTCTCTACTAGGAGTATGCCTCATAGTACAAATCCTCACAGGCTTATTCCTAGCANNNNNNNNNNNNNNNNNNNNNNTAACAGCATTTTCAT
+SRR2190795.2 HWI-ST1147:240:C5NY7ACXX:1:1101:1598:2247 length=100
@@@7D8+@A:1CFG<C:23<:E<;FF<BHIIEHG:?:??CDF<9DCGGG?1?FEG@@<@CA#######################################
@SRR2190795.3 HWI-ST1147:240:C5NY7ACXX:1:1101:1641:2250 length=100
AGAAGGTCTTAGATCAGAAGGAGCACAGACTGGATGGTCGTGTCATTGACCCTAAAAAGGCTANNNNNNNNNNNNNNNNNNNNNTGAAGAAAATCTTTGT
+SRR2190795.3 HWI-ST1147:240:C5NY7ACXX:1:1101:1641:2250 length=100
BC@FFFDDHHHHHJJJJJJJJJJJJJJJJJJJJIJJJFHGHHEGHIIIHJIJJIJJIJIJJID#####################################
@SRR2190795.4 HWI-ST1147:240:C5NY7ACXX:1:1101:1851:2233 length=100
GGGATTTCATGGCCTCCACGTAATTATTGGCTCAACTTTCCTAATTGTCTGTCTACTACGACANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTNNCNNN
+SRR2190795.4 HWI-ST1147:240:C5NY7ACXX:1:1101:1851:2233 length=100
@@?DDBDDFFDDDGHGGGGI?B;FFHGHA@FEHGHDDGHEGGFGHIGEHIIHIGGBGACD6AH#####################################
@SRR2190795.5 HWI-ST1147:240:C5NY7ACXX:1:1101:1957:2243 length=100
CAGCCATTGTGGCTCCCGATGGCTTTGACATCATTGACATGACAGCCGGAGGTCAGATAAACTNNNNNNNNNNNNNNNNNNNNNNATCNGTGGCAAAGGT
+SRR2190795.5 HWI-ST1147:240:C5NY7ACXX:1:1101:1957:2243 length=100
@CCFFFFFHHHHAHJJJIJJJJJJIJJIGGGIFIJIIHIIGGJJJJJJJFHIGIJHHHHHHFC#####################################
```
# 4. è´¨é‡æ§åˆ¶
## 4.1 è´¨é‡è¯„ä¼°
æ‹¿åˆ°æµ‹åºæ•°æ®æ–‡ä»¶ï¼Œåœ¨åºåˆ—æ¯”å¯¹ä¹‹å‰éœ€è¦å¯¹æµ‹åºæ–‡ä»¶çš„æµ‹åºè´¨é‡è¿›è¡ŒæŸ¥çœ‹ï¼Œå› ä¸ºä¸åŒæµ‹åºæ•°æ®æ¥æºæµ‹åºè´¨é‡ä¹Ÿä¸ä¸€æ ·ï¼Œä¸ºäº†ä¿è¯åç»­åˆ†æçš„æœ‰æ•ˆæ€§å’Œå¯é æ€§ï¼Œéœ€è¦å¯¹è´¨é‡è¿›è¡Œè¯„ä¼°ï¼Œå¦‚æœæ•°æ®å¾ˆå·®é‚£ä¹ˆåœ¨åç»­åˆ†æçš„æ—¶å€™å°±éœ€è¦æ³¨æ„äº†ã€‚è¿™é‡Œä½¿ç”¨fastqcè¿›è¡Œè´¨é‡è¯„ä¼°
```
ç”¨æ³•
fastqc [é€‰é¡¹] [æµ‹åºæ–‡ä»¶]

å®é™…ä½¿ç”¨
cd /mnt/c/shengxin/project/rat/sequence
# å› ä¸ºç¨‹åºä¸ä¼šè‡ªåŠ¨æ–°å»ºç›®å½•ï¼Œè¿™é‡Œæ–°å»ºä¸€ä¸ªç›®å½•
mkdir -p ../output/fastqc

# -t æŒ‡å®šçº¿ç¨‹æ•°
# -o æŒ‡å®šè¾“å‡ºæ–‡ä»¶å¤¹
# *.gz è¡¨ç¤ºè¿™ä¸ªç›®å½•ä¸‹ä»¥ .gz çš„æ‰€æœ‰æ–‡ä»¶
fastqc -t 6 -o ../output/fastqc *.gz
```
è¿è¡Œè¿‡ç¨‹ä¸­ä¼šå‡ºç°åˆ†æçš„è¿›ç¨‹ï¼Œåœ¨åˆ†æå®Œæˆä¹‹åä¼šæœ‰åˆ†ææŠ¥å‘Šç”Ÿæˆã€‚
```
cd /mnt/c/shengxin/project/rat/output/fastqc
ls
SRR2190795_fastqc.html  SRR2240182_fastqc.zip   SRR2240184_fastqc.html  SRR2240185_fastqc.zip   SRR2240187_fastqc.html  SRR2240228_fastqc.zip
SRR2190795_fastqc.zip   SRR2240183_fastqc.html  SRR2240184_fastqc.zip   SRR2240186_fastqc.html  SRR2240187_fastqc.zip
SRR2240182_fastqc.html  SRR2240183_fastqc.zip   SRR2240185_fastqc.html  SRR2240186_fastqc.zip   SRR2240228_fastqc.html
```
è¿™é‡Œçš„.htmlç”¨æµè§ˆå™¨æ‰“å¼€ï¼ŒæŸ¥çœ‹ä¸€ä¸‹æƒ…å†µï¼Œ

å¯ä»¥çœ‹åˆ°è¿™ä¸ªæµ‹åºè´¨é‡ä¸æ˜¯ç‰¹åˆ«å¥½ï¼Œ

æœ‰å…³fastqçš„æŠ¥å‘Šè§£è¯»ï¼Œè¿™é‡Œæœ‰ä¸€ç¯‡æ–‡ç« å†™çš„å¯ä»¥ç”¨FastQCæ£€æŸ¥äºŒä»£æµ‹åºåŸå§‹æ•°æ®çš„è´¨é‡[https://www.plob.org/article/5987.html]

ç»¿è‰²è¡¨ç¤ºé€šè¿‡ï¼Œçº¢è‰²è¡¨ç¤ºæœªé€šè¿‡ï¼Œé»„è‰²è¡¨ç¤ºä¸å¤ªå¥½ã€‚ä¸€èˆ¬è€Œè¨€RNA-Seqæ•°æ®åœ¨sequence deplication levels æœªé€šè¿‡æ˜¯æ¯”è¾ƒæ­£å¸¸çš„ã€‚æ¯•ç«Ÿä¸€ä¸ªåŸºå› ä¼šå¤§é‡è¡¨è¾¾ï¼Œä¼šæµ‹åˆ°å¾ˆå¤šé

è¿™é‡Œå› ä¸ºæœ‰å¤šä»½æŠ¥å‘Šï¼Œæœ‰æ—¶å€™æŸ¥çœ‹ä¸æ˜¯ç‰¹åˆ«æ–¹ä¾¿ï¼Œè¿™é‡Œæœ‰ä¸€ä¸ªå°†æ‰€æœ‰çš„```fastqc```çš„æ£€æµ‹æŠ¥å‘Šåˆå¹¶åˆ°ä¸€ä¸ªæ–‡ä»¶ä¸Šçš„ç¨‹åº```multiqc```
```
cd /mnt/c/shengxin/project/rat/output/fastqc

multiqc .
```
è¿›ç¨‹
```
  /// MultiQC ğŸ” | v1.14

|           multiqc | Search path : /mnt/c/shengxin/project/rat/output/fastqc
|         searching | â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â” 100% 16/16
|            fastqc | Found 8 reports
|           multiqc | Compressing plot data
|           multiqc | Report      : multiqc_report.html
|           multiqc | Data        : multiqc_data
|           multiqc | MultiQC complete
```
## 4.2 å‰”é™¤æ¥å¤´ä»¥åŠæµ‹åºè´¨é‡å·®çš„ç¢±åŸº

ä¸Šé¢çœ‹åˆ°ï¼Œåœ¨æ¥å¤´é‚£é‡Œæ˜¯æ˜¾ç¤ºçš„é€šè¿‡ï¼Œä½†æ˜¯å¯ä»¥çœ‹åˆ°æœ‰éƒ¨åˆ†æ˜¯æœ‰4ä¸ªç¢±åŸºä¸æ¥å¤´åºåˆ—åŒ¹é…çš„ï¼Œå±äºIlluminaçš„é€šç”¨æ¥å¤´ã€‚å¦å¤–ä¹Ÿå¯ä»¥çœ‹åˆ°ï¼Œé™¤äº†å¯èƒ½å­˜åœ¨æ¥å¤´çš„æƒ…å†µï¼Œåœ¨æµ‹åºè´¨é‡é‚£é‡Œä¹Ÿå¯ä»¥çœ‹åˆ°åœ¨5'ç«¯å­˜åœ¨ä½è´¨é‡çš„æµ‹åºåŒºåŸŸï¼Œæ‰€ä»¥åƒä¸¤ç«¯è¿™ç§ä½è´¨é‡çš„åŒºåŸŸä¹Ÿæ˜¯è¦å»é™¤çš„çš„ï¼Œè¿™ä¸€æ­¥é‡‡ç”¨trimmomaticè¿›è¡Œã€‚

```
cd /mnt/c/shengxin/project/rat/sequence
# æ–°å»ºæ–‡ä»¶å¤¹
mkdir -p ../output/adapter/

# å¾ªç¯å¤„ç†æ–‡ä»¶å¤¹ä¸‹çš„
for i in $(ls *.fastq.gz);
do
    cutadapt -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT \
    --minimum-length 30 --overlap 4 --trim-n \
    -o ../output/adapter/${i}  ${i}
done

    #for in å¾ªç¯ç»“æ„
    # --minimum-length å¦‚æœå‰”é™¤æ¥å¤´åreadé•¿åº¦ä½äº30ï¼Œè¿™æ¡readå°†ä¼šè¢«ä¸¢å¼ƒ
    # --overlap        å¦‚æœä¸¤ç«¯çš„åºåˆ—ä¸æ¥å¤´æœ‰4ä¸ªç¢±åŸºçš„åŒ¹é…å°†ä¼šè¢«å‰”é™¤
    # --trim-n         å‰”é™¤ä¸¤ç«¯çš„N
    # -a å»é™¤3ç«¯å¼•ç‰©åºåˆ—
    #AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCTæ˜¯Illuminaçš„å°RNAæµ‹åºadapterï¼ˆIllumina Paried End PCR Primer 1ï¼‰
```
## 4.4 å†æ¬¡å»é™¤ä½è´¨é‡åŒºåŸŸ
```
cd /mnt/c/shengxin/project/rat/output/adapter/
mkdir ../trim

$ parallel -j 4 "
  # LEADING:20ï¼Œä»åºåˆ—çš„å¼€å¤´å¼€å§‹å»æ‰è´¨é‡å€¼å°äº 20 çš„ç¢±åŸº
  # TRAILING:20ï¼Œä»åºåˆ—çš„æœ«å°¾å¼€å§‹å»æ‰è´¨é‡å€¼å°äº 20 çš„ç¢±åŸº
  # SLIDINGWINDOW:5:15ï¼Œä» 5' ç«¯å¼€å§‹ä»¥ 5bp çš„çª—å£è®¡ç®—ç¢±åŸºå¹³å‡è´¨é‡ï¼Œå¦‚æœæ­¤å¹³å‡å€¼ä½äº 15ï¼Œåˆ™ä»è¿™ä¸ªä½ç½®æˆªæ–­read
  # Trimmomatic-0.38æˆ‘ä¸‹è½½åˆ°äº†åˆ«çš„ç›˜å†…ï¼Œè®°å¾—æ›´æ”¹è·¯å¾„
  # MINLEN:36ï¼Œ å¦‚æœ reads é•¿åº¦å°äº 30bp åˆ™æ‰”æ‰æ•´æ¡ readã€‚
  java -jar mnt/c/shengxin/biosoft/Trimmomatic-0.38/trimmomatic-0.38.jar \
    SE -phred33 {1} ../trim/{1} \
    LEADING:20 TRAILING:20 SLIDINGWINDOW:5:15 MINLEN:30 \
" ::: $( ls *.gz)

[Trimmomaticå‚æ•°è¯´æ˜](http://www.chenlianfu.com/?p=1948)

# æœ¬å‘½ä»¤è¯´æ˜
#-jar æ‰§è¡Œå°è£…åœ¨ JAR å­˜æ¡£ä¸­çš„ç¨‹åº
java -jar  [Trimmomaticè½¯ä»¶å­˜å‚¨ä½ç½®]\   
   [å•ç«¯æµ‹åº] SE -phred33 {å˜é‡åç§°}  [æŒ‡æ˜å­˜å‚¨è·¯å¾„å’Œåç§°] ../trim/{å˜é‡åç§°}\
   Trimmomaticè½¯ä»¶çš„å‘½ä»¤é€‰é¡¹  
```
ç»“æœï¼š # Trimmomaticè½¯ä»¶å¤„ç†è½¬å…¥åå°ï¼Œä¸ä¼šæ˜¾ç¤ºå®æ—¶æ­¥éª¤ï¼Œè€å¿ƒç­‰å¾…
```
TrimmomaticSE: Started with arguments:
 -phred33 SRR2190795.fastq.gz ../trim/SRR2190795.fastq.gz LEADING:20 TRAILING:20 SLIDINGWINDOW:5:15 MINLEN:30
Automatically using 1 threads
Input Reads: 15105698 Surviving: 14998487 (99.29%) Dropped: 107211 (0.71%)
TrimmomaticSE: Completed successfully
TrimmomaticSE: Started with arguments:
 -phred33 SRR2240183.fastq.gz ../trim/SRR2240183.fastq.gz LEADING:20 TRAILING:20 SLIDINGWINDOW:5:15 MINLEN:30
Automatically using 1 threads
Input Reads: 17621681 Surviving: 17484940 (99.22%) Dropped: 136741 (0.78%)
TrimmomaticSE: Completed successfully
TrimmomaticSE: Started with arguments:
 -phred33 SRR2240184.fastq.gz ../trim/SRR2240184.fastq.gz LEADING:20 TRAILING:20 SLIDINGWINDOW:5:15 MINLEN:30
Automatically using 1 threads
Input Reads: 19777686 Surviving: 19615933 (99.18%) Dropped: 161753 (0.82%)
TrimmomaticSE: Completed successfully
å‡ºç°è¿™ä¸ªè¡¨ç¤ºè¿è¡Œç»“æŸ
```
## 4.3 å†æ¬¡æŸ¥çœ‹è´¨é‡æƒ…å†µ
```
$ cd /mnt/c/shengxin/project/rat/output/trim
#æŒ‚è½½åˆ°å­˜å‚¨gzæ–‡ä»¶çš„æ–‡ä»¶å¤¹ä¸‹
$ mkdir ../fastqc_trim
$ parallel -j 4 "
    fastqc -t 4 -o ../fastqc_trim {1}
" ::: $(ls *.gz)

$ cd ../fastqc_trim
$ multiqc .
```
# 5. å»é™¤rRNAåºåˆ—[å¯ä¸åš]
å¦‚æœåœ¨æå–RNAè¿‡ç¨‹ä¸­æ²¡æœ‰å¯¹RNAè¿›è¡Œç­›é€‰çš„æƒ…å†µä¸‹ï¼Œé‚£ä¹ˆå¾—åˆ°çš„å¤§éƒ¨åˆ†å°†ä¼šæ˜¯rRNAï¼Œè¿™ä¸ªå¯¹äºåç»­çš„åˆ†æå¯èƒ½ä¼šå­˜åœ¨å½±å“ï¼Œå¦å¤–ä¹Ÿä¼šè®©æ¯”å¯¹çš„æ—¶é—´å˜é•¿ã€‚

æ³¨æ„ï¼šåœ¨ä½¿ç”¨sortmernaçš„æ—¶å€™éœ€è¦ç¡®ä¿æµ‹åºæ–‡ä»¶æ˜¯æœªå‹ç¼©çš„æ–‡ä»¶

æ²¡åšï¼ˆï¼‰
# 6. åºåˆ—æ¯”å¯¹
å¾—åˆ°å¹²å‡€çš„æµ‹åºæ•°æ®ä¹‹åï¼Œæ¥å°±å¯ä»¥è¿›è¡Œåºåˆ—æ¯”å¯¹äº†ï¼Œæ¯”å¯¹çš„è¿‡ç¨‹æ˜¯ä¸€ç§å¯»æ‰¾çš„è¿‡ç¨‹ï¼Œå°†readå®šä½åˆ°å®ƒä½äºåŸºå› ç»„çš„ä½ç½®ï¼Œé€šè¿‡æ‰¾å¯»ä¹‹åæ‰èƒ½è¯´è¿™æ¡readæ˜¯å±äºå“ªæ¡åŸºå› çš„ï¼Œè¿™æ ·æ‰èƒ½å¯¹åŸºå› çš„è¡¨è¾¾è¿›è¡Œå®šé‡ã€‚å¦å¤–```RNA-seq```çš„åºåˆ—ä¸åŸºå› ç»„çš„åºåˆ—æœ‰æ—¶å€™ä¼šä¸ä¸€æ ·ï¼Œå› ä¸ºå­˜åœ¨å†…å«å­ä¸å¤–æ˜¾å­è¿™ç§åºåˆ—çš„å·®åˆ«ï¼Œè€Œ```RNA-seq```æ˜¯æµ‹çš„RNAçš„åºåˆ—ï¼Œæ‰€ä»¥ä¼šå‡ºç°è·¨èŒƒå›´çš„åºåˆ—çš„æ¯”å¯¹

## 6.1 å»ºç«‹ç´¢å¼•
è¿™ä¸€æ­¥ä½¿ç”¨```hisat2```ä¸­çš„å·¥å…·```hisat2-build```å»ºç«‹ç´¢å¼•ã€‚

- ç”¨æ³•

```hisat2-build [é€‰é¡¹] [åŸºå› ç»„åºåˆ—(.fa)] [ç´¢å¼•æ–‡ä»¶çš„å‰ç¼€å]```
- å¼€å§‹ä½¿ç”¨
```
$ cd /mnt/c/shengxin/project/rat/genome
$ mkdir index
$ cd index

$ hisat2-build  -p 6 ../R7.2.chr1.fa R7.2.chr1
å¸ˆå…„åªç”¨ä¸€å·æŸ“è‰²ä½“åšäº†æ¯”å¯¹ï¼Œå‘ç°ç»“æœä¸å¤ªå¯¹ï¼Œæ‰€ä»¥æ¥ä¸‹æ¥ä»¥å…¨åŸºå› ç»„åšæ¯”å¯¹
hisat2-build  -p 6 ../R7.2.fa R7.2
#p è¿è¡Œçº¿ç¨‹æ•°
```
åœ¨è¿è¡Œè¿‡ç¨‹ä¸­ä¼šæœ‰éƒ¨åˆ†ä¿¡æ¯æç¤ºï¼Œå…¶ä¸­è¯´åˆ°å»ºç«‹ç´¢å¼•æ–‡ä»¶çš„åˆ†å—æƒ…å†µä»¥åŠè¿è¡Œæ—¶é—´çš„ç»Ÿè®¡

ç´¢å¼•å»ºç«‹å®Œæˆä¹‹ååœ¨```c/mnt/c/shengxin/project/rat/genome```æ–‡ä»¶å¤¹ä¸‹ä¼šå‡ºç°

ç»“æœ:
```
ls
R7.2.1.ht2  R7.2.2.ht2  R7.2.3.ht2  R7.2.4.ht2  R7.2.5.ht2  R7.2.6.ht2  R7.2.7.ht2  R7.2.8.ht2

```
8ä¸ªæ–‡ä»¶ï¼Œè¿™äº›æ–‡ä»¶æ˜¯å¯¹åŸºå› ç»„è¿›è¡Œå‹ç¼©ä¹‹åçš„æ–‡ä»¶ï¼Œè¿™ä¸ªå°†åŸºå› ç»„åºåˆ—æ•°æ®åˆ†å—æˆäº†8ä»½ï¼Œåœ¨æ‰§è¡Œåºåˆ—æ¯”å¯¹çš„æ—¶å€™ç›´æ¥ä½¿ç”¨è¿™äº›æ–‡ä»¶è€Œä¸æ˜¯åŸºå› ç»„R7.2.faæ–‡ä»¶
# 6.2 å¼€å§‹æ¯”å¯¹
è¿™é‡Œä½¿ç”¨hasat2è¿›è¡Œæ¯”å¯¹
```
ç”¨æ³•
hisat2 [é€‰é¡¹] -x [ç´¢å¼•æ–‡ä»¶] [ -1 1æµ‹åºæ–‡ä»¶ -2 2æµ‹åºæ–‡ä»¶ -U æœªæˆå¯¹æµ‹åºæ–‡ä»¶ ] [ -S è¾“å‡ºçš„samæ–‡ä»¶ ]
å®é™…ä½¿ç”¨
cd /mnt/c/shengxin/project/rat/output

mkdir align
cd trim  # æ²¡æœ‰ç»è¿‡rRNAçš„å»é™¤ï¼Œæ‰€ä»¥å’Œå¸ˆå…„çš„ä¸ä¸€æ ·

parallel -k -j 4 "    #-k --keep-order å¼ºåˆ¶ä½¿è¾“å‡ºä¸å‚æ•°ä¿æŒé¡ºåºï¼›-j å¹¶è¡Œä»»åŠ¡æ•°
    hisat2 -t -x ../../genome/index/R7.2 \    # -t å‚æ•°è¡¨ç¤ºè¾“å‡ºåŸºäºæ–‡æœ¬çš„ SAM æ ¼å¼;-x å®šä½åˆ°ç´¢å¼•æ–‡ä»¶
      -U {1} -S ../align/{1}.sam \#-U {1}.fq.gz å‚æ•°æ˜¯è¦æ¯”å¯¹çš„è¾“å…¥æ–‡ä»¶ï¼Œå…¶ä¸­ {1} æ˜¯ä»å‚æ•°ä¸­è·å–çš„æ–‡ä»¶åï¼ˆä¸å¸¦æ‰©å±•åï¼‰'-S ../align/{1}.sam å‚æ•°æŒ‡å®šè¾“å‡ºçš„ SAM æ–‡ä»¶åï¼Œæ ¼å¼ä¸º "../align/æ–‡ä»¶å.sam"
      # -U <r>ï¼šå•ç«¯æµ‹åºlistï¼Œè‹¥ä¸ºlistï¼Œä½¿ç”¨é€—å·éš”å¼€ï¼Œ-U lane1.fq,lane2.fq,lane3.fq,lane4.fq
      # -S <hit> ï¼šSAMå†™å…¥çš„æ–‡ä»¶åï¼Œé»˜è®¤å†™å…¥åˆ°æ ‡å‡†è¾“å‡ºä¸­

      2>../align/{1}.log      #å°†æ ‡å‡†é”™è¯¯è¾“å‡ºé‡å®šå‘åˆ° "../align/æ–‡ä»¶å.log" æ–‡ä»¶ä¸­
" ::: $(ls *.gz | perl -p -e 's/.fq.gz$//')    # perl è¯­å¥èµ·åˆ°æ‰¹é‡æ›¿æ¢ä½œç”¨ï¼Œå¾—åˆ°ä¸å¸¦æ‰©å±•åçš„æ–‡ä»¶å
#:::å°† perl è„šæœ¬è¾“å‡ºçš„ç»“æœä½œä¸ºå‚æ•°ä¼ é€’ç»™ parallel å‘½ä»¤

è¿è¡Œä¸­å‡ºç°hisat2åšæ¯”å¯¹æ—¶å‡ºç°(ERR): hisat2-align exited with value 137 Killed
åŸå› ï¼šå†…å­˜ä¸å¤Ÿï¼Œå°†è¿è¡Œçº¿ç¨‹æ”¹æˆ1
```
æ¯”å¯¹å®Œæˆä¹‹åå¯ä»¥è¿›å…¥æ–‡ä»¶å¤¹æŸ¥çœ‹ä¸€ä¸‹æ—¥å¿—ä¿¡æ¯
```
cd /mnt/c/shengxin/project/rat/output/align

cat SRR2190795.fastq.gz.log

Time loading forward index: 00:00:44
Time loading reference: 00:00:01
Multiseed full-index search: 00:07:27
14998487 reads; of these:
  14998487 (100.00%) were unpaired; of these:
    1198616 (7.99%) aligned 0 times
    13113773 (87.43%) aligned exactly 1 time
    686098 (4.57%) aligned >1 times
92.01% overall alignment rate
Time searching: 00:07:29
Overall time: 00:08:13

```

- æ€»ç»“æ¯”å¯¹æƒ…å†µ
  
è‡ªå·±å†™ä¸€ä¸ªè„šæœ¬å°†åºåˆ—æ¯”å¯¹ç‡å’Œæ—¶é—´è¿›è¡Œç»Ÿè®¡
```
cd /mnt/c/shengxin/project/rat/output/align
file_list=($(ls *.log))

echo -e "sample\tratio\ttime"# è¾“å‡ºè¡¨å¤´ï¼ŒåŒ…æ‹¬ "sample"ã€"ratio" å’Œ "time" ä¸‰åˆ—
for i in ${file_list[@]};
do
    prefix=$(echo ${i} | perl -p -e 's/\.log//')# æå–å‡ºæ¯ä¸ªæ–‡ä»¶çš„å‰ç¼€ï¼ˆå»æ‰ ".log" åç¼€ï¼‰
    echo -n -e "${prefix}\t"#è¾“å‡ºå‰ç¼€ï¼Œå¹¶ä½¿ç”¨ -n å‚æ•°ä½¿å¾—ä¸‹ä¸€æ¡å‘½ä»¤è¾“å‡ºä¸æ¢è¡Œã€‚è¾“å‡ºæ ¼å¼ä¸ºå‰ç¼€åŠ ä¸€ä¸ªåˆ¶è¡¨ç¬¦ã€‚
    cat ${i} |
      grep -E "(overall alignment rate)|(Overall time)" |#-Eå°†æ ·å¼ä¸ºå»¶ä¼¸çš„æ­£åˆ™è¡¨è¾¾å¼æ¥ä½¿ç”¨
      #ç”¨ grep è¿›è¡Œæ­£åˆ™åŒ¹é…ï¼Œåªä¿ç•™åŒ…å« "overall alignment rate" æˆ– "Overall time" çš„è¡Œ
      perl -n -e '
        if(m/alignment/){
          $hash{precent} = $1 if m/([\d.]+)%/;#å¦‚æœå½“å‰è¡ŒåŒ…å« "alignment"ï¼Œåˆ™ä½¿ç”¨æ­£åˆ™è¡¨è¾¾å¼ m/([\d.]+)%/ åŒ¹é…å‡ºç™¾åˆ†æ¯”æ•°å€¼ï¼Œå¹¶å°†å…¶å­˜å…¥å“ˆå¸Œè¡¨ä¸­çš„ "precent" é”®
        }elsif(m/time/){
          if(m/(\d\d):(\d\d):(\d\d)/){
            my $time = $1 * 60 + $2 + $3 / 60;
            $hash{time} = $time;#å¦‚æœå½“å‰è¡ŒåŒ…å« "time"ï¼Œåˆ™ä½¿ç”¨æ­£åˆ™è¡¨è¾¾å¼ m/(\d\d):(\d\d):(\d\d)/ æå–å‡ºæ—¶é—´å€¼ï¼Œå¹¶å°†å…¶è½¬æ¢ä¸ºåˆ†é’Ÿæ•°ï¼Œå­˜å…¥å“ˆå¸Œè¡¨ä¸­çš„ "time" é”®
          }
        }
        END{
          $hash{precent} = "NA" if not exists $hash{precent};
          $hash{time} = "NA" if not exists $hash{time};#åœ¨ END å—ä¸­åˆ¤æ–­è‹¥å“ˆå¸Œè¡¨ä¸­ä¸å­˜åœ¨ "precent" æˆ– "time" é”®ï¼Œåˆ™å°†ç›¸åº”çš„å€¼è®¾ä¸º "NA
          printf "%.2f\t%.2f\n", $hash{precent}, $hash{time};#printf æ˜¯æŒ‡æ ¼å¼åŒ–è¾“å‡ºå‡½æ•°ï¼Œä¸»è¦åŠŸèƒ½æ˜¯å‘æ ‡å‡†è¾“å‡ºè®¾å¤‡æŒ‰è§„å®šæ ¼å¼è¾“å‡ºä¿¡æ¯;æŒ‰ç…§ "%.2f\t%.2f" çš„æ ¼å¼è¾“å‡ºã€‚è¿™é‡Œçš„ "%.2f" è¡¨ç¤ºæµ®ç‚¹æ•°ä¿ç•™ä¸¤ä½å°æ•°ã€‚
        }
      '
done
```
- æ ¼å¼è½¬åŒ–ä¸æ’åº

SAMæ ¼å¼æ˜¯ç›®å‰ç”¨æ¥å­˜æ”¾å¤§é‡æ ¸é…¸æ¯”å¯¹ç»“æœä¿¡æ¯çš„é€šç”¨æ ¼å¼ï¼Œä¹Ÿæ˜¯äººç±»èƒ½å¤Ÿâ€œç›´æ¥â€é˜…è¯»çš„æ ¼å¼ç±»å‹ï¼Œè€ŒBAMå’ŒCRAMæ˜¯ä¸ºäº†æ–¹ä¾¿ä¼ è¾“ï¼Œé™ä½å­˜å‚¨å‹åŠ›å°†SAMè¿›è¡Œå‹ç¼©å¾—åˆ°çš„æ ¼å¼å½¢å¼ã€‚

```
$ cd /mnt/c/shengxin/project/rat/output/align
$ parallel -k -j 4 "
    samtools sort -@ 4 {1}.sam > {1}.sort.bam     
    # å°†æ–‡ä»¶å¤¹å†…samæ–‡ä»¶å…¨éƒ¨è½¬æ¢ä¸ºå…¶äºŒè¿›åˆ¶bamæ–‡ä»¶ä»¥å‡å°‘å†…å­˜,å¹¶æŒ‰ç…§ read name æ’åº
    #-@ 4è¡¨ç¤ºä½¿ç”¨4ä¸ªçº¿ç¨‹æ¥åŠ å¿«å¤„ç†é€Ÿåº¦
    
    
    # samtools indexå¯¹æ’åºå¥½çš„bamæ–‡ä»¶ï¼Œå¯ä»¥é€šè¿‡ä»¥ä¸‹å‘½ä»¤è¿›è¡Œindex
    ï¼ˆæ³¨æ„åªèƒ½å¯¹æ’åºè¿‡çš„æ–‡ä»¶è¿›è¡Œindexï¼‰ï¼Œé»˜è®¤åœ¨å½“å‰æ–‡ä»¶å¤¹äº§ç”Ÿ*.baiçš„indexæ–‡ä»¶
    samtools index {1}.sort.bam

" ::: $(ls *.sam | perl -p -e 's/\.sam$//')
# è¯¥æ­¥äº§ç”Ÿ sort.bam å’Œ .sort.bam.bai æ ¼å¼çš„æ–‡ä»¶
è¡¥å……
1. samtools sort [options] input.bam  

options:
-n : æ ¹æ®readçš„nameè¿›è¡Œæ’åºï¼Œé»˜è®¤å¯¹æœ€å·¦ä¾§åæ ‡è¿›è¡Œæ’åº
-o : è®¾ç½®æ’åºåè¾“å‡ºæ–‡ä»¶çš„æ–‡ä»¶å
-O : åè·Ÿsamæˆ–bamï¼Œè§„å®šæ’åºåè¾“å‡ºæ–‡ä»¶çš„æ ¼å¼ï¼Œé»˜è®¤æ˜¯bam
-@ : åè·Ÿæ­£æ•´æ•°ï¼ŒæŒ‡å®šåˆ†ææ‰€ç”¨çº¿ç¨‹æ•°
-m : åè·Ÿæ•°å­— + K/M/Gï¼ŒæŒ‡å®šæ¯ä¸ªçº¿ç¨‹æ‰€ä½¿ç”¨å†…å­˜é‡  

$ rm *.sam

$ ls
$ rm *.sam
                                                      
$ ls
samtools.756.3746.tmp.0002.bam  samtools.786.3758.tmp.0003.bam    SRR2240182.fastq.gz.log           SRR2240184.fastq.gz.sort.bam.bai  SRR2240187.fastq.gz.sort.bam
samtools.756.3746.tmp.0003.bam  samtools.865.3236.tmp.0000.bam    SRR2240182.fastq.gz.sort.bam      SRR2240185.fastq.gz.log           SRR2240187.fastq.gz.sort.bam.bai
samtools.757.3459.tmp.0002.bam  samtools.865.3236.tmp.0001.bam    SRR2240182.fastq.gz.sort.bam.bai  SRR2240185.fastq.gz.sort.bam      SRR2240228.fastq.gz.log
samtools.757.3459.tmp.0003.bam  samtools.866.1878.tmp.0000.bam    SRR2240183.fastq.gz.log           SRR2240185.fastq.gz.sort.bam.bai  SRR2240228.fastq.gz.sort.bam
samtools.779.2110.tmp.0000.bam  samtools.866.1878.tmp.0001.bam    SRR2240183.fastq.gz.sort.bam      SRR2240186.fastq.gz.log           SRR2240228.fastq.gz.sort.bam.bai
samtools.779.2110.tmp.0002.bam  SRR2190795.fastq.gz.log           SRR2240183.fastq.gz.sort.bam.bai  SRR2240186.fastq.gz.sort.bam
samtools.779.2110.tmp.0003.bam  SRR2190795.fastq.gz.sort.bam      SRR2240184.fastq.gz.log           SRR2240186.fastq.gz.sort.bam.bai
samtools.786.3758.tmp.0002.bam  SRR2190795.fastq.gz.sort.bam.bai  SRR2240184.fastq.gz.sort.bam      SRR2240187.fastq.gz.log
```
# 7. è¡¨è¾¾é‡ç»Ÿè®¡
ä½¿ç”¨HTSEQ-count - htseqçš„ä½¿ç”¨æ–¹æ³•å’Œè®¡ç®—åŸç†[https://htseq.readthedocs.io/en/master/count.html#]

å¦‚ä½•åˆ¤æ–­ä¸€ä¸ª reads å±äºæŸä¸ªåŸºå› ï¼Œ htseq-count æä¾›äº† union, intersection_strict,intersection_nonempty 3 ç§æ¨¡å‹ï¼Œå¦‚å›¾ï¼ˆå¤§å¤šæ•°æƒ…å†µä¸‹ä½œè€…æ¨èç”¨ union æ¨¡å‹ï¼‰ï¼Œå®ƒæè¿°äº†åœ¨å¤šç§æƒ…å†µä¸‹ï¼Œæ¯”å¯¹åˆ°åŸºå› ç»„ä¸Šçš„readåˆ†é…çš„é—®é¢˜ï¼Œåœ¨è¿™äº›é—®é¢˜ä¸­ï¼Œæœ€éš¾åˆ†é…çš„å°±æ˜¯ä¸€æ¡readåœ¨ä¸¤ä¸ªåŸºå› ç›¸äº¤çš„åœ°æ–¹æ¯”å¯¹ä¸Šäº†ä¹‹åçš„æƒ…å†µã€‚ä¸€èˆ¬æƒ…å†µä¸‹ä½œè€…æ¨èä½¿ç”¨unionçš„æ–¹å¼ã€‚
```                       
ç”¨æ³•
htseq-count [options] <alignment_files> <gff_file  >
```
å‚æ•°è¯´æ˜    
| å‚æ•° | è¯´æ˜ | 
| :----:| :----:| 
|-f --format	 | default: sam è®¾ç½®è¾“å…¥æ–‡ä»¶çš„æ ¼å¼ï¼Œè¯¥å‚æ•°çš„å€¼å¯ä»¥æ˜¯samæˆ–bamã€‚ | 
|-r --order | default: name è®¾ç½®samæˆ–bamæ–‡ä»¶çš„æ’åºæ–¹å¼ï¼Œè¯¥å‚æ•°çš„å€¼å¯ä»¥æ˜¯nameæˆ–posã€‚å‰è€…è¡¨ç¤ºæŒ‰readåè¿›è¡Œæ’åºï¼Œåè€…è¡¨ç¤ºæŒ‰æ¯”å¯¹çš„å‚è€ƒåŸºå› ç»„ä½ç½®è¿›è¡Œæ’åºã€‚è‹¥æµ‹åºæ•°æ®æ˜¯åŒæœ«ç«¯æµ‹åºï¼Œå½“è¾“å…¥sam/bamæ–‡ä»¶æ˜¯æŒ‰posæ–¹å¼æ’åºçš„æ—¶å€™ï¼Œä¸¤ç«¯readsçš„æ¯”å¯¹ç»“æœåœ¨sam/bamæ–‡ä»¶ä¸­ä¸€èˆ¬ä¸æ˜¯ç´§é‚»çš„ä¸¤è¡Œï¼Œç¨‹åºä¼šå°†readså¯¹çš„ç¬¬ä¸€ä¸ªæ¯”å¯¹ç»“æœæ”¾å…¥å†…å­˜ï¼Œç›´åˆ°è¯»å–åˆ°å¦ä¸€ç«¯readçš„æ¯”å¯¹ç»“æœã€‚å› æ­¤ï¼Œé€‰æ‹©poså¯èƒ½ä¼šå¯¼è‡´ç¨‹åºä½¿ç”¨è¾ƒå¤šçš„å†…å­˜ï¼Œå®ƒä¹Ÿé€‚åˆäºæœªæ’åºçš„sam/bamæ–‡ä»¶ã€‚è€Œposæ’åºåˆ™è¡¨ç¤ºç¨‹åºè®¤ä¸ºåŒæœ«ç«¯æµ‹åºçš„readsæ¯”å¯¹ç»“æœåœ¨ç´§é‚»çš„ä¸¤è¡Œä¸Šï¼Œä¹Ÿé€‚åˆäºå•ç«¯æµ‹åºçš„æ¯”å¯¹ç»“æœã€‚å¾ˆå¤šå…¶å®ƒè¡¨è¾¾é‡åˆ†æè½¯ä»¶è¦æ±‚è¾“å…¥çš„sam/bamæ–‡ä»¶æ˜¯æŒ‰posæ’åºçš„ï¼Œä½†HTSeqæ¨èä½¿ç”¨nameæ’åºï¼Œä¸”ä¸€èˆ¬æ¯”å¯¹è½¯ä»¶çš„é»˜è®¤è¾“å‡ºç»“æœä¹Ÿæ˜¯æŒ‰nameè¿›è¡Œæ’åºçš„ã€‚ | 
| -s --stranded | 	default: yes è®¾ç½®æ˜¯å¦æ˜¯é“¾ç‰¹å¼‚æ€§æµ‹åºã€‚è¯¥å‚æ•°çš„å€¼å¯ä»¥æ˜¯yes,noæˆ–reverseã€‚noè¡¨ç¤ºéé“¾ç‰¹å¼‚æ€§æµ‹åºï¼›è‹¥æ˜¯å•ç«¯æµ‹åºï¼Œyesè¡¨ç¤ºreadæ¯”å¯¹åˆ°äº†åŸºå› çš„æ­£ä¹‰é“¾ä¸Šï¼›è‹¥æ˜¯åŒæœ«ç«¯æµ‹åºï¼Œyesè¡¨ç¤ºread1æ¯”å¯¹åˆ°äº†åŸºå› æ­£ä¹‰é“¾ä¸Šï¼Œread2æ¯”å¯¹åˆ°åŸºå› è´Ÿä¹‰é“¾ä¸Šï¼›reverseè¡¨ç¤ºåŒæœ«ç«¯æµ‹åºæƒ…å†µä¸‹ä¸yeså€¼ç›¸åçš„ç»“æœã€‚æ ¹æ®è¯´æ˜æ–‡ä»¶çš„ç†è§£ï¼Œä¸€èˆ¬æƒ…å†µä¸‹åŒæœ«ç«¯é“¾ç‰¹å¼‚æ€§æµ‹åºï¼Œè¯¥å‚æ•°çš„å€¼åº”è¯¥é€‰æ‹©reverseï¼ˆæœ¬äººæš‚æ—¶æ²¡æœ‰æµ‹è¯•è¯¥å‚æ•°ï¼‰ã€‚ | 
|-a --a	 |default: 10 å¿½ç•¥æ¯”å¯¹è´¨é‡ä½äºæ­¤å€¼çš„æ¯”å¯¹ç»“æœã€‚åœ¨0.5.4ç‰ˆæœ¬ä»¥å‰è¯¥å‚æ•°é»˜è®¤å€¼æ˜¯0ã€‚ |
| -t --type |	default: exon ç¨‹åºä¼šå¯¹è¯¥æŒ‡å®šçš„featureï¼ˆgtf/gffæ–‡ä»¶ç¬¬ä¸‰åˆ—ï¼‰è¿›è¡Œè¡¨è¾¾é‡è®¡ç®—ï¼Œè€Œgtf/gffæ–‡ä»¶ä¸­å…¶å®ƒçš„featureéƒ½ä¼šè¢«å¿½ç•¥ã€‚ |
|-i --idattr| default: gene_id è®¾ç½®feature IDæ˜¯ç”±gtf/gffæ–‡ä»¶ç¬¬9åˆ—é‚£ä¸ªæ ‡ç­¾å†³å®šçš„ï¼›è‹¥gtf/gffæ–‡ä»¶å¤šè¡Œå…·æœ‰ç›¸åŒçš„feature IDï¼Œåˆ™å®ƒä»¬æ¥è‡ªåŒä¸€ä¸ªfeatureï¼Œç¨‹åºä¼šè®¡ç®—è¿™äº›featuresçš„è¡¨è¾¾é‡ä¹‹å’Œèµ‹ç»™ç›¸åº”çš„feature IDã€‚|
| -m --mode	| default: union è®¾ç½®è¡¨è¾¾é‡è®¡ç®—æ¨¡å¼ã€‚è¯¥å‚æ•°çš„å€¼å¯ä»¥æœ‰union, intersection-strict and intersection-nonemptyã€‚è¿™ä¸‰ç§æ¨¡å¼çš„é€‰æ‹©è¯·è§ä¸Šé¢å¯¹è¿™3ç§æ¨¡å¼çš„ç¤ºæ„å›¾ã€‚ä»å›¾ä¸­å¯çŸ¥ï¼Œå¯¹äºåŸæ ¸ç”Ÿç‰©ï¼Œæ¨èä½¿ç”¨intersection-strictæ¨¡å¼ï¼›å¯¹äºçœŸæ ¸ç”Ÿç‰©ï¼Œæ¨èä½¿ç”¨unionæ¨¡å¼ã€‚|
| -o --samout	 | è¾“å‡ºä¸€ä¸ªsamæ–‡ä»¶ï¼Œè¯¥samæ–‡ä»¶çš„æ¯”å¯¹ç»“æœä¸­å¤šäº†ä¸€ä¸ªXFæ ‡ç­¾ï¼Œè¡¨ç¤ºè¯¥readæ¯”å¯¹åˆ°äº†æŸä¸ªfeatureä¸Šã€‚ |
| -q --quiet | ä¸è¾“å‡ºç¨‹åºè¿è¡Œçš„çŠ¶æ€ä¿¡æ¯å’Œè­¦å‘Šä¿¡æ¯ã€‚|
| -h --help	|è¾“å‡ºå¸®åŠ©ä¿¡æ¯ã€‚ |	

```
cd /mnt/c/shengxin/project/rat/output
mkdir HTseq

cd align
parallel -j 1 "
    htseq-count -s no -f bam {1}.sort.bam ../../annotation/R7.2.gtf \#-s noè¡¨ç¤ºéé“¾ç‰¹å¼‚æ€§æµ‹åº;-f bam bamæ ¼å¼ï¼›
      >../HTseq/{1}.count  2>../HTseq/{1}.log
" ::: $(ls *.fastq.gz.sort.bam | perl -p -e 's/\.sort\.bam$//')
```

æŸ¥çœ‹ç”Ÿæˆçš„æ–‡ä»¶
```
cd /mnt/c/shengxin/project/rat/output/HTseq
cat SRR2190795.fastq.gz.count | head -n 10
```
ç»“æœï¼ˆç¬¬ä¸€åˆ—ï¼šåŸºå› çš„IDï¼Œç¬¬äºŒåˆ—ï¼šreadè®¡æ•°ï¼‰
```
ENSRNOG00000000001      2
ENSRNOG00000000007      3
ENSRNOG00000000008      0
ENSRNOG00000000009      0
ENSRNOG00000000010      0
ENSRNOG00000000012      0
ENSRNOG00000000017      10
ENSRNOG00000000021      0
ENSRNOG00000000024      842
ENSRNOG00000000033      28
```

# 8. åˆå¹¶è¡¨è¾¾çŸ©é˜µä¸æ ‡å‡†åŒ–
## 8.1 åˆå¹¶
è¿™é‡Œå°±æ˜¯å°†ä¸‹é¢çš„è¿™ç§è¡¨åˆå¹¶ä¸ºä¸€å¼ è¡¨ï¼Œä½œä¸ºä¸€ä¸ªæ•´ä½“è¾“å…¥åˆ°åç»­åˆ†æçš„ç¨‹åºä¸­
```
      æ ·æœ¬1 |        æ ·æœ¬2 |       æ ·æœ¬3
åŸºå› 1   x   | åŸºå› 1    x   | åŸºå› 1   x
åŸºå› 2   x   | åŸºå› 2    x   | åŸºå› 2   x
åŸºå› 3   x   | åŸºå› 3    x   | åŸºå› 3   x
åŸºå› 4   x   | åŸºå› 4    x   | åŸºå› 4   x
```
åˆå¹¶ä¸º
```
      æ ·æœ¬1   æ ·æœ¬2  æ ·æœ¬3
åŸºå› 1   x      x      x
åŸºå› 2   x      x      x
åŸºå› 3   x      x      x
åŸºå› 4   x      x      x
```
ä¸‹é¢ä½¿ç”¨Rè¯­è¨€ä¸­çš„mergeå°†è¡¨æ ¼åˆå¹¶
```
rm(list=ls())#æ¸…é™¤Rç¯å¢ƒä¸­çš„æ‰€æœ‰å¯¹è±¡ï¼Œç¡®ä¿å¼€å§‹æ—¶æ²¡æœ‰ä»»ä½•å˜é‡å’Œæ•°æ®
setwd("~/project/rat/output/HTseq")#è®¾ç½®å½“å‰å·¥ä½œç›®å½•

# å¾—åˆ°æ–‡ä»¶æ ·æœ¬ç¼–å·
files <- list.files(".", "*.count")#è·å–å½“å‰å·¥ä½œç›®å½•ä¸‹æ‰€æœ‰ä»¥ ".count" ç»“å°¾çš„æ–‡ä»¶
f_lists <- list() #åˆ›å»ºä¸€ä¸ªç©ºçš„åˆ—è¡¨ f_lists ç”¨äºå­˜å‚¨æ–‡ä»¶å¯¹åº”çš„å‰ç¼€å’Œæ–‡ä»¶å
for(i in files){
    prefix = gsub("(_\\w+)?\\.count", "", i, perl=TRUE)#gsub ï¼ˆ"æ›¿æ¢å¯¹è±¡â€ï¼Œâ€œæ›¿æ¢æˆä»€ä¹ˆå†…å®¹â€ï¼Œå“ªä¸€åˆ—ï¼‰ gsubä¸åé¢çš„æ‹¬å·ä¹‹é—´ä¸èƒ½æœ‰ç©ºæ ¼
#ä»æ–‡ä»¶åä¸­æå–å‡ºå‰ç¼€éƒ¨åˆ†ï¼ˆå»æ‰ ".count" åç¼€ï¼‰ã€‚å°†æå–çš„å‰ç¼€èµ‹å€¼ç»™å˜é‡ prefix
    f_lists[[prefix]] = i#å°†æ–‡ä»¶åï¼ˆåŒ…æ‹¬æ‰©å±•åï¼‰å­˜å‚¨åœ¨ f_lists åˆ—è¡¨ä¸­ï¼Œä»¥å‰ç¼€ä¸ºç´¢å¼•
}


id_list <- names(f_lists)#è·å– f_lists åˆ—è¡¨çš„ç´¢å¼•ï¼ˆå‰ç¼€ï¼‰ï¼Œå°†å…¶å­˜å‚¨åœ¨ id_list ä¸­
data <- list()#åˆ›å»ºä¸€ä¸ªç©ºçš„åˆ—è¡¨ data ç”¨äºå­˜å‚¨è¯»å–çš„æ–‡ä»¶å†…å®¹
count <- 0#åˆå§‹åŒ–è®¡æ•°å™¨ count ä¸º0ï¼Œç”¨äºè®°å½•å½“å‰å¤„ç†çš„æ–‡ä»¶æ•°é‡
for(i in id_list){
  count <- count + 1#è®¡æ•°å™¨ count åŠ 1ï¼Œè¡¨ç¤ºå¼€å§‹å¤„ç†ä¸€ä¸ªæ–°çš„æ–‡ä»¶
  a <- read.table(f_lists[[i]], sep="\t", col.names = c("gene_id",i))#ä»æ–‡ä»¶ f_lists[[i]] ä¸­è¯»å–æ•°æ®ï¼Œä½¿ç”¨åˆ¶è¡¨ç¬¦"\t"ä½œä¸ºåˆ†éš”ç¬¦ï¼Œå¹¶æŒ‡å®šåˆ—åä¸º "gene_id" å’Œå˜é‡ iã€‚
  data[[count]] <- a#å°†è¯»å–çš„æ•°æ®å­˜å‚¨åœ¨ data åˆ—è¡¨ä¸­ï¼Œåˆ—è¡¨ç´¢å¼•ä¸º count
}

# åˆå¹¶æ–‡ä»¶
data_merge <- data[[1]]#å°†ç¬¬ä¸€ä¸ªæ–‡ä»¶çš„æ•°æ®èµ‹å€¼ç»™ data_mergeï¼Œä½œä¸ºåç»­åˆå¹¶çš„åŸºç¡€
for(i in seq(2, length(id_list))){#ä»ç¬¬äºŒä¸ªæ–‡ä»¶å¼€å§‹ï¼Œéå†å‰©ä½™çš„æ–‡ä»¶
    data_merge <- merge(data_merge, data[[i]],by="gene_id")
}
#ä½¿ç”¨ merge å‡½æ•°é€ä¸ªå°†æ–‡ä»¶çš„æ•°æ®æŒ‰ç…§ "gene_id" åˆ—è¿›è¡Œåˆå¹¶ï¼Œå°†ç»“æœèµ‹å€¼ç»™ data_merge
write.csv(data_merge, "merge.csv", quote = FALSE, row.names = FALSE)#quoteå‚æ•°è®¾ä¸ºFALSEè¡¨ç¤ºä¸å¯¹æ•°æ®è¿›è¡Œå¼•å·åŒ…å›´ï¼Œrow.nameså‚æ•°è®¾ä¸ºFALSEè¡¨ç¤ºä¸å†™å…¥è¡Œåã€‚
#æœ€åä¼šå¾—åˆ°ä¸€ä¸ªç»Ÿè®¡å¥½çš„excelè¡¨æ ¼
```
# 8.2 æ•°æ®æ ‡å‡†åŒ–
## 8.2.1 ç®€ä»‹
è¡¨è¾¾é‡æ˜¯ä¸ªä»€ä¹ˆï¼Ÿ
åœ¨ä¸€ä¸ªç»†èƒä¸­ï¼Œå¦‚æœç»Ÿè®¡æŸä¸€ä¸ªåŸºå› çš„è¡¨è¾¾å¾—åˆ°çš„æ‰€æœ‰çš„RNAï¼Œè¿™ä¸ªæ•°é‡å°±æ˜¯ç»å¯¹çš„æ•°é‡ï¼Œå°±æ˜¯å®é™…ä¸Šæœ‰å¤šå°‘ã€‚ä½†æ˜¯åœ¨RNA-seqä¸­ï¼Œå¹¶ä¸çŸ¥é“è¢«ç”¨äºæµ‹åºçš„ç»„ç»‡å—æœ‰å¤šå°‘ä¸ªç»†èƒï¼Œå¦å¤–æå–RNAçš„è¿‡ç¨‹ä¸­æŸå¤±äº†å¤šå°‘ï¼Œé‚£ä¹ˆæœ€åæˆ‘ä»¬é€šè¿‡è¿™ä¸ªread countä»¥åŠæ ‡å‡†åŒ–ä¹‹åå¾—åˆ°çš„å¹¶ä¸æ˜¯çœŸå®çš„å…·ä½“çš„æ•°é‡ï¼Œè¿™ä¸ªæ•°é‡æ˜¯ç›¸å¯¹å®šé‡ï¼Œå°±æ˜¯è¿™ä¸ªæ•°å€¼å•ç‹¬æ‹¿å‡ºæ¥æ²¡æœ‰ä»€ä¹ˆæ„ä¹‰ï¼Œä½†æ˜¯åœ¨æ•°æ®ç›¸äº’æ¯”è¾ƒä¸­æ‰æ˜¾ç¤ºå‡ºæ„ä¹‰æ¥ã€‚

read countä¸ç›¸å¯¹è¡¨è¾¾é‡
å¾—åˆ°çš„åŸå§‹read countå¹¶ä¸èƒ½ä½“ç°å‡ºåŸºå› ä¸åŸºå› ä¹‹é—´çš„ç›¸å¯¹çš„è¡¨è¾¾é‡çš„å…³ç³»ã€‚æ¯”å¦‚ç»è¿‡HTseq-countä¹‹åå¾—åˆ°çš„é‚£äº›æ•°å€¼ï¼Œè¿™ä¸ªæ•°å€¼å°±æ˜¯è¯´è½åœ¨åŸºå› åŒºåŸŸå†…çš„readçš„æ•°é‡ã€‚ä½†æ˜¯å¦‚ä¸Šå›¾æ‰€ç¤ºï¼Œä¸åŒçš„åŸºå› çš„é•¿åº¦ä¸åŒï¼Œé‚£ä¹ˆå¯¹åº”çš„readæ¯”å¯¹åˆ°çš„åŒºåŸŸçš„å¤§å°ä¸åŒï¼ŒåŸºå› ä¹‹é—´çš„é•¿åº¦ä¸åŒè¿™å°±å¸¦æ¥äº†ç›´æ¥æ¯”è½åœ¨åŸºå› ä¸Šçš„readæ•°é‡æ¥è¯´æ˜è¡¨è¾¾é‡å°±æ˜¯ä¸å…¬å¹³çš„è¿™ç§æƒ…å†µ Sampleï¼ˆå°±å¥½æ¯”ç›´æ¥æ¯”ä¸¤ä¸ªäººçš„ä½“é‡æ¥åˆ¤å®šèƒ–ç˜¦ä¸€æ ·ï¼Œæ¯”å¦‚ä¸€ä¸ª100æ–¤çš„6-7å²çš„å°èƒ–å­å’Œ110æ–¤çš„æˆå¹´äººä¸€æ ·ï¼Œä¸è€ƒè™‘èº«é«˜å› ç´ è¿™ç§å…³é”®æ˜¯æ²¡æœ‰æ„ä¹‰çš„ï¼‰ï¼Œæ‰€ä»¥éœ€è¦æ ¹æ®åŸºå› çš„é•¿åº¦æ¥å¯¹åŸå§‹çš„read countè¿›è¡Œè½¬åŒ–ä¹‹åæ‰èƒ½å…¬å¹³ã€‚è¿™é‡Œçš„æ ‡å‡†åŒ–æ˜¯å±äºæ ·æœ¬å†…çš„å„ä¸ªåŸºå› ä¹‹é—´çš„è¡¨è¾¾é‡çš„æ ‡å‡†åŒ–ã€‚

ä½†æ˜¯åç»­åªæ˜¯ä¸ºäº†åˆ†æåŸºå› çš„å·®å¼‚è¡¨è¾¾ï¼Œæ‰€ä»¥åœ¨å¯¹æµ‹åºæ·±åº¦è¿›è¡Œæ ‡å‡†åŒ–ä¹‹åå°±å¯ä»¥ç›´æ¥å¯¹ä¸åŒæ ·æœ¬åŒä¸€ä¸ªåŸºå› ä¹‹é—´çš„read countæ•°è¿›è¡Œæ¯”è¾ƒï¼Œå› ä¸ºå¹¶ä¸æ¶‰åŠåˆ°ä¸€ä¸ªæ ·æœ¬å†…ä¸åŒåŸºå› çš„å¯¹æ¯”ã€‚ä¸ä¸Šé¢çš„æ ·æœ¬å†…çš„ä¸åŒï¼Œè¿™ä¸ªæ˜¯å±äºæ ·æœ¬é—´çš„æ ‡å‡†åŒ–ï¼Œå› ä¸ºä¸åŒRNA-seqçš„æµ‹åºæ·±åº¦å¯èƒ½æ˜¯æœ‰å·®åˆ«çš„ã€‚

```            
æ ·æœ¬é—´ç›¸åŒåŸºå› çš„å¯¹æ¯”ï¼ˆTMMåˆ†ä½æ•°æ ‡å‡†åŒ–æˆ–è€…æ·±åº¦æ ‡å‡†åŒ–ï¼Œæˆ–è€…è¿˜æ˜¯CPMã€RPKMã€FPKMã€TPMæ ‡å‡†åŒ–ï¼‰
                    |
          sample1   |   sample2      sample3
gene1       x    <--+-->  x            x
                                       ^
                                       |
                                       +------ æ ·æœ¬å†…çš„ä¸åŒåŸºå› å¯¹æ¯”ï¼ˆRPKMã€FPKMã€TPMæ ‡å‡†åŒ–ï¼‰
                                       |
                                       v
gene2       x             x            x
gene3       x             x            x
```
æ ·æœ¬ä¹‹é—´çš„æ ‡å‡†åŒ–ä¼šä½¿ç”¨åˆ†ä½æ•°æ ‡å‡†åŒ–æˆ–è€…CPM(counts per million)æˆ–è€…log-CPMè¿›è¡Œï¼ˆlog-CPMçš„è®¡ç®—ä¸ºlog2(CPM + 2 * 10^6/N)ï¼Œè¿™æ ·çš„å–å¯¹æ•°é¿å…äº†å¯¹0å–å¯¹æ•°çš„æƒ…å†µï¼Œè¿™ä¸ªè¯´æ˜åœ¨ä¸‹é¢çš„vstæ ‡å‡†åŒ–é‚£é‡Œä¼šæåˆ°ï¼Œä¹‹æ‰€ä»¥å–å¯¹æ•°æ˜¯å› ä¸ºä½¿æ‰€æœ‰æ ·æœ¬é—´çš„logå€æ•°å˜åŒ–ï¼ˆlog-fold-changeï¼‰å‘0æ¨ç§»è€Œå‡å°ä½è¡¨è¾¾åŸºå› é—´å¾®å°è®¡æ•°å˜åŒ–å¸¦æ¥çš„å·¨å¤§çš„ä¼ªå·®å¼‚æ€§ï¼Œå¦‚æœæ€»çš„readæ•°é‡æœ‰4åƒä¸‡ï¼Œé‚£ä¹ˆ0çš„å€¼å°±æ˜¯log2(2 / 40) = -4.32ï¼‰

å› ä¸ºè¿™å‡ ä¸ªè®¡ç®—æ–¹æ³•å¹¶ä¸æ¶‰åŠåˆ°åŸºå› é•¿åº¦ï¼Œæ‰€ä»¥åœ¨è®¡ç®—ä¸Šæ˜¯ç›¸å¯¹æ–¹ä¾¿çš„ï¼Œå½“ç„¶äº†ï¼Œå¦‚æœç ”ç©¶çš„æ ·æœ¬åœ¨å¯å˜å‰ªæ¥çš„ä½¿ç”¨ä¸Šæœ‰è¾ƒå¤§å·®å¼‚ï¼Œé‚£ä¹ˆåœ¨æ¯”è¾ƒçš„æ—¶å€™ä½¿ç”¨ä¸Šé¢çš„å‡ ç§æ–¹å¼å¯èƒ½å°±ä¸å¥½ï¼Œè¿™ä¸ªæ—¶å€™éœ€è¦è€ƒè™‘é•¿åº¦çš„å› ç´ äº†ã€‚

ä¸ºäº†åç»­å¯èƒ½éœ€è¦çš„QPCRå®éªŒéªŒè¯ï¼Œè¿™é‡Œå°†æ•°æ®è¿›è¡Œä¸€ä¸ªæ ·æœ¬å†…çš„æ ‡å‡†åŒ–çš„è®¡ç®—ã€‚ä½†æ˜¯è¿™ä¸ªæ•°å€¼ä¸ç”¨äºåç»­çš„å·®å¼‚åˆ†æå½“ä¸­ã€‚ç›¸å…³åšæ–‡RNA-Seqåˆ†æ|RPKM, FPKM, TPM, å‚»å‚»åˆ†ä¸æ¸…æ¥šï¼Ÿï¼›BBQ(ç”Ÿç‰©ä¿¡æ¯åŸºç¡€é—®é¢˜35ï¼Œ36)ï¼šRNA-Seq æ•°æ®çš„å®šé‡ä¹‹RPKMï¼ŒFPKMå’ŒTPMï¼Œä½†æ˜¯ç›®å‰å­˜åœ¨äº‰è®®ç©¶ç«Ÿæ˜¯ä½¿ç”¨FPKMè¿˜æ˜¯TPMçš„é—®é¢˜ï¼Œè¿™é‡Œå¯¹ä¸¤ç§æ–¹æ³•éƒ½è¿›è¡Œè®¡ç®—ã€‚

## 8.2.2 cufflinks
å¯¹äºTophatçš„æ¯”å¯¹ç»“æœï¼Œå¯ä»¥ä½¿ç”¨Cufflinksè½¯ä»¶è¿›è¡Œåç»­çš„åˆ†æã€‚è€ŒCufflinksè¿™ä¸ªè½¯ä»¶å…¶å®æ˜¯ä¸ªå¥—è£…ï¼Œå…¶ä¸­æ˜¯åŒ…å«è®¸å¤šçš„è½¯ä»¶ï¼ŒåŒ…æ‹¬cuffflinksï¼Œcuffmergeï¼Œcuffnormï¼Œcuffdiffç­‰ï¼Œå¯ä»¥å¯¹è½¬å½•æœ¬è¿›è¡Œæ‹¼æ¥ã€æ•´åˆã€è¡¨è¾¾ä¸°åº¦çš„ä¼°è®¡ä»¥åŠå·®å¼‚è¡¨è¾¾çš„ç­›é€‰ã€‚

## 8.3æ‰‹åŠ¨è®¡ç®—
http://www.360doc.com/content/18/0112/02/50153987_721216719.shtml è¿™ç¯‡æ–‡ç« è®²çš„å¾ˆæ¸…æ¥š

å·²çŸ¥ï¼š æ¯ä¸ªæ ·æœ¬ä¸Šä¸åŒåŸºå› å¯¹åº”åˆ°çš„readsæ•°ç›®
å¯è®¡ç®—ï¼š æ¯ä¸ªæ ·æœ¬çš„æ€»readsæ•°ç›®
éœ€è¦é¢å¤–è·å¾—çš„æ•°æ®ï¼š åŸºå› é•¿åº¦

é¦–å…ˆå¾—åˆ°ç›¸å…³åŸºå› çš„é•¿åº¦ä¿¡æ¯
```
#ä½¿ç”¨RåŒ… GenomicFeatures ç»Ÿè®¡æ¯ä¸ªåŸºå› çš„å¤–æ˜¾å­é•¿åº¦
library(GenomicFeatures)
# æ„å»ºGrangeså¯¹è±¡
txdb <- makeTxDbFromGFF("r7.2.gff" )
# æŸ¥æ‰¾åŸºå› çš„å¤–æ˜¾å­
exons_gene <- exonsBy(txdb, by = "gene")
# è®¡ç®—æ€»é•¿åº¦
# reduce()ã€width()æ˜¯Irangeå¯¹è±¡çš„æ–¹æ³•
gene_len <- list()
for(i in names(exons_gene)){
    range_info = reduce(exons_gene[[i]])
    width_info = width(range_info)
    sum_len    = sum(width_info)
    gene_len[[i]] = sum_len
}

# æˆ–è€…å†™ä¸ºlapplyçš„å½¢å¼(å¿«å¾ˆå¤š)
gene_len <- lapply(exons_gene,function(x){sum(width(reduce(x)))})

data <- t(as.data.frame(gene_len))
# å†™å…¥æ–‡ä»¶
write.table(data, file = "r7.2_gene_len.tsv", row.names = TRUE, sep="\t", quote = FALSE, col.names = FALSE)

#æœ€åä¼šå¾—åˆ°ä¸€ä¸ªæ–‡æœ¬æ–‡ä»¶ï¼Œç”¨è®°äº‹æœ¬æ‰“å¼€
è¦æ³¨æ„è¿™é‡Œçš„å•ä½æ˜¯kbï¼Œä¸æ˜¯bp
```
å¼€å§‹è®¡ç®—RPKM å’Œ TPM
```
cpmè®¡ç®—å…¬å¼

CPM = (10^6 * nr) / N
CPM : Counts per million
nr : æ¯”å¯¹è‡³ç›®æ ‡åŸºå› çš„readæ•°é‡
N : æ˜¯æ€»æœ‰æ•ˆæ¯”å¯¹è‡³åŸºå› ç»„çš„readæ•°é‡

ç”¨é€”âˆ¶åœ¨æŸäº›æƒ…å†µä¸‹ï¼Œåªæƒ³äº†è§£æ¯ä¸ªåŸºå› è¢«è¦†ç›–åˆ°çš„ç›¸å¯¹readsæ•°ï¼Œè€Œä¸å¸Œæœ›å¯¹å…¶åšé•¿åº¦æ ¡æ­£ï¼Œå°±ä¼šä½¿ç”¨è¿™ä¸ªæŒ‡æ ‡ã€‚

ç”¨æ€»readsè¿›è¡Œå‡ä¸€åŒ–æ˜¯æœ€ç®€å•çš„æ–¹æ³•ï¼Œå…¶åŸºäºä»¥ä¸‹ä¸¤ä¸ªåŸºæœ¬å‡è®¾âˆ¶
1ï¼‰ç»å¤§å¤šæ•°çš„geneè¡¨è¾¾é‡ä¸å˜ï¼›
2ï¼‰é«˜è¡¨è¾¾é‡çš„geneè¡¨è¾¾é‡ä¸å‘ç”Ÿæ”¹å˜ï¼›

RPKMè®¡ç®—å…¬å¼ RPKMé€‚ç”¨äºå•ç«¯æµ‹åºæ–‡åº“RPKMé€‚ç”¨äºå•ç«¯æµ‹åºæ–‡åº“

RPKM = (10^6 * nr) / (L * N)
RPKM: Reads Per Kilobase per Million
nr : æ¯”å¯¹è‡³ç›®æ ‡åŸºå› çš„readæ•°é‡
L : ç›®æ ‡åŸºå› çš„å¤–æ˜¾å­é•¿åº¦ä¹‹å’Œé™¤ä»¥1000(å› æ­¤ï¼Œè¦æ³¨æ„è¿™é‡Œçš„Lå•ä½æ˜¯kbï¼Œä¸æ˜¯bp)
N : æ˜¯æ€»æœ‰æ•ˆæ¯”å¯¹è‡³åŸºå› ç»„çš„readæ•°é‡

ç”¨é€”âˆ¶ç”¨äºä¸åŸºå› è¡¨è¾¾é‡ç›¸å…³çš„åæœŸåˆ†æã€‚åŸºå› è¡¨è¾¾è¶‹åŠ¿åˆ†æã€WGCNAå…±è¡¨è¾¾ç½‘ç»œæ„å»ºï¼Œçƒ­å›¾ç»˜åˆ¶ç­‰éƒ½ä½¿ç”¨ã€‚
```
```
#!R
# =========== RPKM =============

gene_len_file <- "rn6_gene_len.tsv"
count_file <- "samples.count"

gene_len <- read.table(gene_len_file, header = FALSE, row.name = 1)
colnames(gene_len) <- c("length")

count <- read.table(count_file, header = FALSE, row.name = 1)
colnames(count) <- c("count")
# all read number
all_count <- sum(count["count"])

RPKM <- c()
for(i in row.names(count)){
    count_ = 0
    exon_kb = 1
    rpkm = 0
    count_ = count[i, ]
    exon_kb  = gene_len[i, ] / 1000
    rpkm    = (10 ^ 6 * count_ ) / (exon_kb * all_count )
    RPKM = c(RPKM, rpkm)
}
```
```
TPMè®¡ç®—å…¬å¼

TPM = nr * read_r * 10^6 / g_r * T
T   = âˆ‘(ni * read_i / g_i)
ç®€è¨€ä¹‹

TPM = (nr / g_r) * 10^6 / âˆ‘(ni / gi)
TPM : Transcripts Per Million
nr : æ¯”å¯¹è‡³ç›®æ ‡åŸºå› çš„readæ•°é‡
read_r: æ˜¯æ¯”å¯¹è‡³åŸºå› rçš„å¹³å‡readé•¿åº¦
g_r : æ˜¯åŸºå› rçš„å¤–æ˜¾å­é•¿åº¦ä¹‹å’Œï¼ˆè¿™é‡Œæ— éœ€å°†å…¶é™¤ä»¥1000ï¼‰
```
```
# =========== è®¡ç®—TPM ============
# é¦–å…ˆå¾—åˆ°æ€»çš„ç»“æœ
sum_ <- 0
for(i in row.names(count)){
    count_ = 0
    exon = 1
    count_ = count[i, ]
    exon  = gene_len[i, ]
    value = count_ / exon
    if(is.na(value)){
        print(paste(i, " is error! please check"))
    }else{
        sum_ = sum_ + value
    }
}

TPM <- c()
for(i in row.names(count)){
    count_ = 0
    exon = 1
    count_ = count[i, ]
    exon  = gene_len[i, ]
    tpm = (10 ^ 6 * count_ / exon ) / sum_     
    # å…ˆé™¤ä»¥åŸºå› é•¿åº¦å†é™¤ä»¥æ€»readæ•°
    TPM = c(TPM, tpm)
}

count["RPKM"] <- RPKM
count["TPM"] <- TPM       
           
write.table(count, "123.normalize.count", col.names = TRUE, row.names = TRUE, sep="\t", quote = FALSE)
```
å®é™…ä¸Šä¸Šé¢ä¸¤ç§è®¡ç®—æ–¹æ³•éƒ½é»˜è®¤äº†æ¯ä¸€ä¸ªåŸºå› çš„æ‰€æœ‰çš„å¤–æ˜¾å­éƒ½ä¼šè¡¨è¾¾å‡ºæ¥ï¼Œä½†æ˜¯å®é™…ä¸Šå¯èƒ½å¹¶ä¸æ˜¯çš„ï¼Œæ‰€ä»¥å¦‚æœæƒ³è¦è¾ƒä¸ºç²¾ç¡®çš„å€¼å°±éœ€è¦å¯¹bamæ–‡ä»¶åŠ ä¸Šgffæ³¨é‡Šæ–‡ä»¶è¿›è¡Œè”åˆä¹‹åå¾—åˆ°çœŸå®çš„åŸºå› çš„å¤–æ˜¾å­çš„é•¿åº¦ã€‚ä½†æ˜¯è¿™ä¸ªå¯¹äºRPKMçš„æ•´ä½“è®¡ç®—æ¥è¯´æ²¡æœ‰ä»€ä¹ˆå½±å“ï¼Œä½†æ˜¯å¯¹äºTPMæ¥è¯´å¯èƒ½ä¼šæœ‰å¾®å¼±çš„å·®å¼‚ã€‚

ä¸Šé¢å¯ä»¥çœ‹åˆ°ã€‚RPKMçš„æ˜¯å•ä¸ªåŸºå› çš„read countæ•°åšåˆ†å­ï¼Œç„¶ååˆ†æ¯å‡å’Œæ€»readæ•°ç›¸å…³ï¼Œå®é™…ä¸Šè¿™æ ·åœ¨ä¸€å®šç¨‹åº¦ä¸Šæ¶ˆé™¤äº†æµ‹åºçš„å»ºåº“çš„å¤§å°çš„å·®å¼‚ã€‚ä½†æ˜¯è¿™ä¼šå¸¦æ¥ä¸€å®šçš„é—®é¢˜ï¼Œå°±æ˜¯ä¸èƒ½ä¿è¯ä¸åŒçš„æ ·æœ¬ä¸­æ€»çš„RNAçš„è¡¨è¾¾æ€»é‡æ€»æ˜¯ä¸€è‡´çš„ã€‚å‡å¦‚è‚ç»†èƒæ¯”çº¢ç»†èƒçš„RNAæ€»é‡é«˜ï¼Œä½†æ˜¯åœ¨ç»è¿‡RPKMçš„æ—¶å€™å°†
```
æŸåŸºå› è¡¨è¾¾é‡ï¼š 
            +----+ å æ¯”0.01
è‚ç»†èƒæ€»RNAï¼š
            +------------------+


æŸåŸºå› è¡¨è¾¾é‡ï¼š 
            +--+ å æ¯”0.01
çº¢ç»†èƒæ€»RNAï¼š 
            +-------------+
            ```
ç»è¿‡nr / Nå½’ä¸€åŒ–å‹ç¼©åˆ°0~1çš„èŒƒå›´å†…ï¼ˆè¿™ä¸ªæ—¶å€™æ²¡æœ‰ä¹˜ä»¥10^6 / Lè¿™ä¸ªå¸¸é‡ï¼‰ï¼Œé‚£ä¹ˆæŒ‰ç…§æ¯”ä¾‹æ¥è¯´ä¸€æ ·ï¼Œä½†æ˜¯å®é™…ä¸Šçš„RNAè¡¨è¾¾é‡æ•°å€¼æ˜¯ä¸ç­‰çš„ã€‚åªèƒ½è¯´è¡¨è¾¾çš„å æ¯”ç›¸ç­‰ã€‚
```
                    å½’ä¸€åŒ–

----------------------------------------------

è‚ç»†èƒæŸåŸºå› è¡¨è¾¾é‡ï¼š         çº¢ç»†èƒæŸåŸºå› è¡¨è¾¾é‡ï¼š

y^                         y^
1|-------------+           1|-------------+
 |             |            |             |
 |             |            |             |
 |   *         |            |   *         |
 | *           |            | *           |
 +--------------->          +--------------->
               1 x                         1x
               ```
ä½†æ˜¯è¿™ä¸ªæ•°å€¼ç›¸ç­‰äº†ï¼Œæ˜¯å¦èƒ½è¯„åˆ¤ä¸åŒç»„ç»‡ä¸­çš„å…·ä½“çš„åŸºå› è¡¨è¾¾é‡å‘¢ï¼Ÿ

# 9. å·®å¼‚è¡¨è¾¾åˆ†æ

æŸ¥çœ‹å‡ ä¸ªç®¡å®¶åŸºå› çš„è¡¨è¾¾é‡æƒ…å†µã€‚
```GAPDH(ENSRNOG00000018630)ã€beta-actin(ENSRNOG00000034254)```
è¡¥å……ï¼šç®¡å®¶åŸºå› ï¼ˆhousekeeping geneï¼‰ï¼ŒåˆåæŒå®¶åŸºå› ã€çœ‹å®¶åŸºå› æˆ–å®¶åŠ¡åŸºå› ï¼Œæ˜¯æŒ‡åœ¨ç”Ÿç‰©ä½“å†…æ‰€æœ‰ç»†èƒä¸­éƒ½è¡¨è¾¾ï¼Œå¹¶ä¸”ä¸ºç»´æŒç»†èƒåŸºæœ¬ç”Ÿå‘½æ´»åŠ¨æ‰€éœ€è€Œæ—¶åˆ»éƒ½åœ¨è¡¨è¾¾çš„é«˜åº¦ä¿å®ˆçš„åŸºå› ã€‚
```
cd c/shengxin/project/rat/output/HTseq

cat merge.csv | grep -E "ENSRNOG00000018630|ENSRNOG00000034254"
#CSVï¼ˆComma-Separated Valuesï¼Œé€—å·åˆ†éš”çš„å€¼ï¼‰æ˜¯ç”¨äºå­˜å‚¨å’Œè¡¨ç¤ºåŒ…æ‹¬æ–‡æœ¬ã€æ•°å€¼ç­‰å„ç§ç±»å‹çš„æ•°æ®;ç‰¹ç‚¹æ˜¯ï¼šæ–‡ä»¶å†…çš„æ•°æ®ä»¥é€—å· ,åˆ†éš”ï¼Œå‘ˆç°ä¸€ä¸ªè¡¨æ ¼å½¢å¼ã€‚
```
è¿™æ˜¯ä¸¤ä¸ªåŸºå› çš„è¡¨è¾¾é‡æƒ…å†µ
åˆ é™¤HTseq-countçš„æ€»ç»“è¡Œ
```
dataframe <- read.csv("merge.csv", header=TRUE, row.names = 1)
```
åœ¨æ•°æ®ä¸­å­˜åœ¨æ€»ç»“çš„é¡¹ï¼Œè¿™äº›é¡¹å¯¹äºåç»­åˆ†ææœ‰å½±å“ï¼Œåœ¨HTseq-countçš„ç»“æœæœ‰5è¡Œæ€»ç»“çš„å†…å®¹ï¼Œåˆ†åˆ«æ˜¯ï¼š
| é¡¹	| è¯´æ˜ | 
| :----:|:----:| 
|__alignment_not_unique |æ¯”å¯¹åˆ°å¤šä¸ªä½ç½®çš„readsæ•° | 
|__ambiguous | ä¸èƒ½åˆ¤æ–­è½åœ¨é‚£ä¸ªå•ä½ç±»å‹çš„readsæ•°| 
|__no_feature| ä¸èƒ½å¯¹åº”åˆ°ä»»ä½•å•ä½ç±»å‹çš„readsæ•° | 
|__not_aligned |å­˜åœ¨äºSAMæ–‡ä»¶ï¼Œä½†æ²¡æœ‰æ¯”å¯¹ä¸Šçš„readsæ•°| 
|__too_low_aQual	|ä½äº-aè®¾å®šçš„reads mappingè´¨é‡çš„readsæ•°| 


è¿™é‡Œåˆ é™¤æ‰
```
                       SRR2190795 SRR2240182 SRR2240183 SRR2240184
__alignment_not_unique    1237425    1821001    1327114    1554701
__ambiguous                237874     419677     260420     328308
__no_feature              1331291    1927193    1435345    1475574
__not_aligned             1350172    2317888    1262136     963510
__too_low_aQual                 0          0          0          0


                       SRR2240185 SRR2240186 SRR2240187 SRR2240228
__alignment_not_unique    1807092     962538    1703558    1436165
__ambiguous                364825     186332     336135     275106
__no_feature              2239366    1150812    1566965    1425114
__not_aligned             1336922    1027748    1734828    1539988
__too_low_aQual                 0          0          0          0
```
```
# å»é™¤å‰é¢5è¡Œ
countdata <- dataframe[-(1:5),]

# æŸ¥çœ‹æ•°æ®
head(countdata)  

#ä¸ä¼šåœ¨åŸå§‹è¡¨æ ¼ä¸­ä¿®æ”¹ï¼Œåªæ˜¯åœ¨è¿è¡Œè¿‡ç¨‹ä¸­å˜é‡ä¸­ä¿®æ”¹

```
å°†IDçš„ç‰ˆæœ¬å·å»é™¤

æœ‰çš„æ—¶å€™åœ¨åŸºå› ååé¢ä¼šæœ‰.1æˆ–è€….2ç­‰ç­‰çš„æ ‡å·å‡ºç°ï¼ˆè¿™é‡Œæ²¡æœ‰ï¼‰ï¼Œè¿™ä¸ªæ—¶å€™éœ€è¦æŠŠå®ƒé™¤å»

```
# å¾—åˆ°è¡Œçš„å
row_names <- row.names(countdata)

# å¼€å§‹æ›¿æ¢
name_replace <- gsub("\\.\\w+","", row.names(countdata))
# å¥å·.å’ŒåŠ å·w+æ˜¯ç‰¹æ®Šçš„ï¼Œè¦æ·»åŠ \\æ¥è¯†åˆ«
# è¡¨ç¤ºå¥å·åé¢çš„wordéƒ½è¦è¢«æ›¿æ¢ä¸º [ç©º]

row.names(countdata) <- name_replace

```	
- å»é™¤ä½è¡¨è¾¾çš„åŸºå› 
åœ¨ä»»ä½•æ ·æœ¬ä¸­éƒ½æ²¡æœ‰è¶³å¤Ÿå¤šçš„åºåˆ—ç‰‡æ®µçš„åŸºå› åº”è¯¥ä»ä¸‹æ¸¸åˆ†æä¸­è¿‡æ»¤æ‰ã€‚è¿™æ ·åšçš„åŸå› æœ‰å¥½å‡ ä¸ªã€‚ ä»ç”Ÿç‰©å­¦çš„è§’åº¦æ¥çœ‹ï¼Œåœ¨ä»»ä½•æ¡ä»¶ä¸‹çš„è¡¨è¾¾æ°´å¹³éƒ½ä¸å…·æœ‰ç”Ÿç‰©å­¦æ„ä¹‰çš„åŸºå› éƒ½ä¸å€¼å¾—å…³æ³¨ï¼Œå› æ­¤æœ€å¥½å¿½ç•¥ã€‚ ä»ç»Ÿè®¡å­¦çš„è§’åº¦æ¥çœ‹ï¼Œå»é™¤ä½è¡¨è¾¾è®¡æ•°åŸºå› ä½¿æ•°æ®ä¸­çš„å‡å€¼ - æ–¹å·®å…³ç³»å¯ä»¥å¾—åˆ°æ›´ç²¾ç¡®çš„ä¼°è®¡ï¼Œå¹¶ä¸”è¿˜å‡å°‘äº†åœ¨è§‚å¯Ÿå·®å¼‚è¡¨è¾¾çš„ä¸‹æ¸¸åˆ†æä¸­éœ€è¦è¿›è¡Œçš„ç»Ÿè®¡æ£€éªŒçš„æ•°é‡ã€‚
```
countdata <- countdata[rowSums(countdata) > 0,]
```
ä¸æ˜¯åªè¦æ˜¯ä¸ª0çš„æ•°æ®å»åˆ é™¤ï¼Œè€Œæ˜¯æ‰€æœ‰æ ·æœ¬ä¸­è¯¥åŸºå› éƒ½æ˜¯0æ‰åˆ é™¤

åˆ°è¿™é‡Œå°±å¾—åˆ°äº†å¯ä»¥ç”¨äºåç»­å·®å¼‚åˆ†æçš„æ•°æ®äº†

## 9.2 å·®å¼‚åˆ†æ
å·®å¼‚åˆ†æä½¿ç”¨DESeq2åŒ…è¿›è¡Œåˆ†æï¼Œè¿™ä¸ªå¯¹äºè¾“å…¥çš„æ•°æ®æ˜¯åŸå§‹çš„read countï¼Œæ‰€ä»¥ä¸Šè¿°ç»è¿‡HTseqçš„readè®¡æ•°ä¹‹åçš„æ•°æ®å¯ä»¥è¾“å…¥åˆ°DESeq2åŒ…ä¸­è¿›è¡Œå·®å¼‚åˆ†æã€‚å®ƒä¸EdgeRåŒ…ç±»ä¼¼ï¼Œéƒ½æ˜¯åŸºäºè´ŸäºŒé¡¹åˆ†å¸ƒæ¨¡å‹ã€‚åœ¨è½¬å½•ç»„åˆ†æä¸­æœ‰ä¸‰ä¸ªåˆ†æçš„æ°´å¹³åŸºå› æ°´å¹³(gene-level)ã€è½¬å½•æœ¬æ°´å¹³(transcript-level)ã€å¤–æ˜¾å­ä½¿ç”¨æ°´å¹³(exon-usage-level)ã€‚ä½†æ˜¯åŸå§‹çš„read countæ•°é‡å¹¶ä¸èƒ½ä»£è¡¨åŸºå› çš„è¡¨è¾¾é‡ã€‚

è¡¨è¾¾å·®å¼‚åˆ†æåªå¯¹æ¯”ä¸åŒæ ·æœ¬ä¹‹é—´çš„åŒä¸€ä¸ªè½¬å½•æœ¬ï¼Œæ‰€ä»¥ä¸éœ€è¦è€ƒè™‘è½¬å½•æœ¬é•¿åº¦ï¼Œåªè€ƒè™‘æ€»è¯»æ®µæ•°ã€‚ä¸€ä¸ªæœ€ç®€å•æ€æƒ³å°±æ˜¯ï¼Œæ ·æœ¬æµ‹åºå¾—åˆ°çš„æ€»è¯»æ®µæ•°ï¼ˆå®é™…ä¸Šæ˜¯å¯ä»¥æ¯”å¯¹åˆ°è½¬å½•ç»„çš„æ€»è¯»æ®µæ•°ï¼‰è¶Šå¤šï¼Œåˆ™æ¯ä¸ªåŸºå› åˆ†é…åˆ°çš„è¯»æ®µè¶Šå¤šã€‚å› æ­¤æœ€ç®€å•çš„æ ‡å‡†åŒ–å› å­å°±æ˜¯æ€»è¯»æ®µæ•°ï¼Œç”¨æ€»è¯»æ®µæ•°ä½œæ ‡å‡†åŒ–çš„å‰ææ˜¯å¤§éƒ¨åˆ†åŸºå› çš„è¡¨è¾¾æ˜¯éæ˜¾è‘—å˜åŒ–çš„ï¼Œè¿™ä¸åŸºå› èŠ¯ç‰‡ä¸­çš„åŸºæœ¬å‡è®¾ç›¸åŒã€‚ä½†æ˜¯å®é™…å·¥ä½œä¸­å‘ç°å¾ˆå¤šæƒ…å†µä¸‹æ€»è¯»æ®µæ•°ä¸»è¦æ˜¯ä¸€å°éƒ¨åˆ†å¤§é‡è¡¨è¾¾çš„åŸºå› è´¡çŒ®çš„ã€‚Bullardç­‰ï¼ˆ2010ï¼‰åœ¨æ¯”è¾ƒäº†å‡ ç§æ ‡å‡†åŒ–æ–¹æ³•çš„åŸºç¡€ä¸Šå‘ç°åœ¨æ¯ä¸ªæ³³é“å†…ä½¿ç”¨éé›¶è®¡æ•°åˆ†å¸ƒçš„ä¸Šå››åˆ†ä½æ•°ï¼ˆQ75%ï¼‰ä½œä¸ºæ ‡å‡†åŒ–å› å­æ˜¯ä¸€ç§æ›´ç¨³å¥çš„é€‰æ‹©ï¼Œæ€»ä½“è¡¨ç°æ˜¯æ‰€ç ”ç©¶æ–¹æ³•ä¸­æœ€ä¼˜çš„ã€‚

Bioconductorçš„edgeRåŒ…å’ŒDESeqåŒ…åˆ†åˆ«æä¾›äº†ä¸Šå››åˆ†ä½æ•°å’Œä¸­ä½æ•°æ¥ä½œä¸ºæ ‡å‡†åŒ–å› å­ï¼Œå°±æ˜¯å‡ºäºè¿™ä¸ªæ€æƒ³ã€‚Bioconductoråˆ†æRNA-seqæ•°æ®

DESeq2çš„å·®å¼‚åˆ†æçš„æ­¥éª¤
```
æ„å»ºä¸€ä¸ªdds(DESeqDataSet)çš„å¯¹è±¡
åˆ©ç”¨DESeqå‡½æ•°è¿›è¡Œæ ‡å‡†åŒ–å¤„ç†
ç”¨resultå‡½æ•°æ¥æå–å·®å¼‚æ¯”è¾ƒçš„ç»“æœ
```
### 9.2.1 å®‰è£…ä¸åŠ è½½åŒ…
é¦–å…ˆå®‰è£…å¯¹åº”çš„RåŒ…
```
# ä½¿ç”¨bioconductorè¿›è¡Œå®‰è£…
source("http://bioconductor.org/biocLite.R")
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")

# å®‰è£…åŒ…
biocLite("DESeq2")
biocLite("pheatmap")
biocLite("biomaRt")
biocLite("org.Rn.eg.db")
biocLite("clusterProfiler")

# åŠ è½½
library(DESeq2)
library(pheatmap) # ç”¨äºä½œçƒ­å›¾çš„åŒ…
library(biomaRt)
library(org.Rn.eg.db)# å¤§é¼ ï¼Œè¤å®¶é¼ ç‰©ç§åŒ…
library(clusterProfiler) #è¿›è¡ŒGOã€KEGGçš„å¯Œé›†åˆ†æ

#bioMartåŒ…æ˜¯ä¸€ä¸ªè¿æ¥bioMartæ•°æ®åº“çš„Rè¯­è¨€æ¥å£ï¼Œèƒ½é€šè¿‡è¿™ä¸ªè½¯ä»¶åŒ…è‡ªç”±è¿æ¥åˆ°bioMartæ•°æ®åº“
è¿™ä¸ªÂ·åŒ…å¯ä»¥åšä»¥ä¸‹å‡ ä¸ªå·¥ä½œ
1.æŸ¥æ‰¾æŸä¸ªåŸºå› åœ¨æŸ“è‰²ä½“ä¸Šçš„ä½ç½®ã€‚åä¹‹ï¼Œç»™å®šæŸ“è‰²ä½“æ¯ä¸€åŒºé—´ï¼Œè¿”å›è¯¥åŒºé—´çš„åŸºå› sï¼›
2.é€šè¿‡EntrezGeneçš„IDæŸ¥æ‰¾åˆ°ç›¸å…³åºåˆ—çš„GOæ³¨é‡Šã€‚åä¹‹ï¼Œç»™å®šç›¸å…³çš„GOæ³¨é‡Šï¼Œè·å–ç›¸å…³çš„EntrezGeneçš„IDï¼›
3.é€šè¿‡EntrezGeneçš„IDæŸ¥æ‰¾åˆ°ç›¸å…³åºåˆ—çš„ä¸Šæ¸¸100bpåºåˆ—ï¼ˆå¯èƒ½åŒ…å«å¯åŠ¨å­ç­‰è°ƒæ§å…ƒä»¶ï¼‰ï¼›
4.æŸ¥æ‰¾äººç±»æŸ“è‰²ä½“ä¸Šæ¯ä¸€æ®µåŒºåŸŸä¸­å·²çŸ¥çš„SNPsï¼›
5.ç»™å®šä¸€ç»„çš„åºåˆ—IDï¼Œè·å¾—å…¶ä¸­å…·ä½“çš„åºåˆ—ï¼›

# â†‘ å¸ˆå…„çš„ä»£ç ï¼ŒError: With R version 3.5 or greater, install Bioconductor packages using BiocManager; 
see https://bioconductor.org/installï¼Œ
è€Œä¸”package â€˜Biocondutorâ€™ is not available for Bioconductor version '3.15'æ²¡åŠæ³•ç”¨BiocManageræ¥install("Biocondutor")ï¼›
å®˜ç½‘å»ºè®®ï¼Œä¸€äº›bioconductorçš„packagesç”¨managerå®‰è£…å°±å¯ä»¥ï¼Œäºæ˜¯é‡‡ç”¨ BiocManager æ¥å®‰è£…   

æ ¹æ®http://bioconductor.org/install/#bioc-versionä¸­ä»£ç æ¥å®‰è£…

æˆ‘çš„ä»£ç ï¼š

 if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")  

  BiocManager::install("DESeq2")
  BiocManager::install("pheatmap")
  BiocManager::install("biomaRt")
  BiocManager::install("org.Rn.eg.db")
  BiocManager::install("clusterProfiler")


install.packages("BiocManager")
BiocManager::install("DESeq2", force = TRUE)
  BiocManager::install("pheatmap", force = TRUE)
  BiocManager::install("biomaRt", force = TRUE)
  BiocManager::install("org.Rn.eg.db", force = TRUE)
  BiocManager::install("clusterProfiler", force = TRUE)


```
### 9.2.2 æ„å»ºå¯¹è±¡
è¿™é‡Œè¯´ç™½äº†å°±æ˜¯æŠŠæ•°æ®å¯¼å…¥åˆ°Rä¸­ç”Ÿæˆå¯¹åº”çš„æ•°æ®ç»“æ„ï¼Œå®ƒçš„åŸºæœ¬ç”¨æ³•å¦‚ä¸‹ï¼š
```
#å°†åˆ›å»ºçš„å·®å¼‚åˆ†ææ•°æ®é›†å­˜å‚¨åœ¨å˜é‡ dds ä¸­
dds <- DESeqDataSetFromMatrix(countData = cts, colData = coldata, design= ~ batch + condition)
#ç”¨å…¬å¼ "~ batch + condition" è¡¨ç¤ºæ¨¡å‹ä¸­åŒ…æ‹¬æ‰¹æ¬¡æ•ˆåº”ï¼ˆbatchï¼‰å’Œæ¡ä»¶æ•ˆåº”ï¼ˆconditionï¼‰ä¸¤ä¸ªå˜é‡
```
- countDataï¼ˆè¡¨è¾¾çŸ©é˜µï¼‰ï¼šæ˜¯ä¸Šé¢ä¸€æ­¥ç”Ÿæˆçš„ä¸€ä¸ªæ•°æ®æ¡†ï¼ˆåˆ—å¯¹åº”ç€æ¯ä¸€ä¸ªæ ·æœ¬ï¼Œè¡Œå¯¹åº”çš„åŸºå› åç§°ï¼Œä¸­é—´çš„å€¼æ˜¯readçš„è®¡æ•°ï¼‰ï¼Œç±»ä¼¼äºä¸‹é¢çš„
  
| æ ·æœ¬1	 | æ ·æœ¬2 | æ ·æœ¬3 |æ ·æœ¬4|
 :----:|  :----:| :----: |:----: |
| åŸºå› 1	 | 10 |20|15|16|
|åŸºå› 2|0|	0|	2|	2|
|åŸºå› 3|	120|	110|	20	|10|
|åŸºå› 4|	40	|44|	10|	20|
|åŸºå› 5|	20|	10	|13|	12|

- colDataï¼ˆæ ·æœ¬ä¿¡æ¯ï¼‰ï¼šè¿™ä¸ªæ˜¯ç”¨æ¥æè¿°æ ·æœ¬çš„æ˜¯å®éªŒç»„è¿˜æ˜¯å¯¹ç…§ç»„ï¼Œç±»ä¼¼äºä¸‹é¢

|sample|	treatment|
| :----: | :----: |
|Control1|control|
|Control2	|control|
|Experiment1|	treatment|
|Experiment2|	treatment|

treatmentä¸ä¸€å®šå°±æ˜¯æŒ‡ä»£æ ·æœ¬æ˜¯ç»è¿‡ä»€ä¹ˆå¤„ç†çš„ï¼Œä¹Ÿå¯ä»¥æ˜¯ç»†èƒç±»å‹ã€åŸºå› å‹ã€è¡¨ç°å‹ã€æ ·æœ¬å¤„ç†æ–¹å¼ã€æ‰¹æ¬¡ç­‰ç­‰ä¿¡æ¯ï¼Œå› ä¸ºå¦‚æœç›´æ¥ç»™æ ·æœ¬ä¿¡æ¯ç¨‹åºæ˜¯ä¸çŸ¥é“ç©¶ç«Ÿæ˜¯æ€æ ·çš„åˆ†ç»„çš„ï¼Œè€Œè¿™äº›ä¿¡æ¯å°±æ˜¯è¢«ç”¨äºåŒºåˆ†æ ·æœ¬çš„æ€§è´¨å¯¹æ ·æœ¬åˆ†ç»„ï¼Œæ‰€ä»¥è¯´æ˜¯å¾ˆé‡è¦çš„ä¿¡æ¯ï¼Œå¦‚æœåˆ†é”™é‚£ä¹ˆæ•°æ®æ¯”è¾ƒçš„æ—¶å€™å°±ä¼šç›¸åº”çš„å‘ç”Ÿå˜åŒ–ï¼Œæœ€åå¾—åˆ°çš„ç»“æœå°±ä¼šå‘ç”Ÿå˜åŒ–ã€‚	

- designï¼ˆæ ·æœ¬å·®å¼‚æ¯”è¾ƒï¼‰ï¼šå°±æ˜¯æŒ‡å®šæ ·æœ¬ä¾æ®ä»€ä¹ˆåˆ†ä¸ºå®éªŒç»„ä¸å¯¹ç…§ç»„

ä¸Šé¢çš„è¡¨è¾¾çŸ©é˜µå·²ç»å¾—åˆ°äº†ï¼Œä¸‹é¢éœ€è¦ç”Ÿæˆæ ·æœ¬çš„ä¿¡æ¯ï¼Œä¸‹é¢çš„è¡¨æ ¼æˆ‘ç›´æ¥ä»NCBIçš„Run selectorä¸­å¾—åˆ°ã€‚

|Run|BioSample|Sample name|Experiment|LoadDate|MBases	|MBytes|	health state|treatment
| :----: | :----: | :----: | :----: | :----: | :----: |:----: | :----: |:----: |
|SRR2240185|	SAMN03975629|	DEN_1_2__DEN251_255_index7	|SRX1182156	|2015-09-07	|2,195|	1,590	|Liver cirrhosis	|DEN
|SRR2240186	|SAMN03975630|	DEN_4_5__DEN24_59_index12	|SRX1182158	|2015-09-07	|1,128|	815	|Liver cirrhosis|	DEN
|SRR2240187	|SAMN03975631|	PBS_1_2__PBS8_9_index13	|SRX1182166	|2015-09-07	|1,861|	1,342|	Healthy control|	PBS
|SRR2240228	|SAMN03975632|	PBS_3_5__PBS18_19_index14	|SRX1182170	|2015-09-07	|1,649|	1,190	|Healthy control|	PBS

è¿™ä¸ªè¡¨æ ¼è¯´æ˜äº†æ ·æœ¬IDåŠå…¶å¤„ç†çš„æƒ…å†µï¼Œå¯ä»¥çœ‹åˆ°å°±æ˜¯treatmenté‚£ä¸€æ ä¸ä¸€æ ·ï¼Œä¸‹é¢é’ˆå¯¹

è¡¨è¾¾æ•°æ®å·²ç»æœ‰äº†ï¼Œä¸‹é¢æ˜¯å†™ä¸€ä¸‹å®éªŒç»„ä¸å¯¹ç…§ç»„çš„ä¿¡æ¯ï¼Œæ‰“å¼€ç»ˆç«¯ï¼Œcd /mnt/c/shengxin/project/rat
```
cat <<EOF >./phenotype/phenotype.csv
"ids","state","condition","treatment"
"SRR2240185","Liver cirrhosis","DEN","treatment"
"SRR2240186","Liver cirrhosis","DEN","treatment"
"SRR2240187","Healthy control","PBS","control"
"SRR2240228","Healthy control","PBS","control"
EOF
#EOF â€œend of fileâ€ï¼Œè¡¨ç¤ºæ–‡æœ¬ç»“æŸç¬¦
```
ä¸‹é¢å°†è¿™äº›æ•°æ®å¯¼å…¥åˆ°Rä¸­
```
# åˆšæ‰countdataå·²ç»å¾—åˆ°
countdata

# è¯»å–æ ·æœ¬åˆ†ç»„ä¿¡æ¯(æ³¨æ„ï¼Œéœ€è¦åŠ ä¸Šrow.names = 1, header = TRUEï¼Œå°†è¡Œåˆ—åéœ€è¦çœ‹å¥½)
coldata <- read.table("../phenotype/phenotype.csv", row.names = 1, header = TRUE, sep = "," )
# ç¡®è®¤ä¸€ä¸‹è¡Œåˆ—åæ˜¯å¦æœ‰ï¼ˆä¸æ˜¯ç®€å•çš„æ•°å€¼ï¼‰
head(coldata)
# è°ƒæ•´æ•°æ®é¡ºåº
countdata <- countdata[row.names(coldata)]

# æ„å»ºddså¯¹è±¡
dds <- DESeqDataSetFromMatrix(countData = countdata, colData = coldata, design= ~ treatment)

# æŸ¥çœ‹dds
dds
```
### 9.2.3 æ ·æœ¬ç›¸å…³æ€§ --é€‰æ‹©æ€§æ­¥éª¤ï¼Œä¸æ˜¯å·®å¼‚åˆ†ææ‰€å¿…é¡»çš„
å› ä¸ºå­˜åœ¨å¾ˆå¤šåŸºå› çš„å·®åˆ«ç­‰å› ç´ ï¼Œåœ¨æŸäº›åŸºå› ä¸Šå¯èƒ½æ ·æœ¬é—´å‡ ä¹æ²¡æœ‰å·®åˆ«ï¼Œä½†æ˜¯æ€»ä½“æ¥çœ‹å°±ä¼šæœ‰è¾ƒå¤§å·®åˆ«äº†ï¼Œè¿™é‡Œå¯¹åŒ…å«ä¼—å¤šçš„åŸºå› è¿™æ ·çš„å› ç´ çš„æƒ…å†µä¸‹è¿›è¡Œæ ·æœ¬ç›¸å…³æ€§è¿›è¡Œè¯„ä¼°ï¼Œè¯„ä¼°æ ·æœ¬çš„é‡å¤ç»„ä¹‹é—´æ˜¯å¦å¾ˆç›¸ä¼¼æˆ–è€…æ˜¯å¦å®éªŒç»„ä¸å¯¹ç…§ç»„ä¹‹é—´å·®åˆ«æ˜æ˜¾ã€‚

PCAåˆ†æ(principal components analysis)
ç”±äºä¸Šé¢å¾—åˆ°çš„æ˜¯æœ€åŸå§‹çš„read countï¼Œä½†æ˜¯PCAåˆ†æéœ€è¦å¯¹æ•°æ®è¿›è¡Œè½¬åŒ–æ‰èƒ½è¿›è¡Œã€‚ä¸€èˆ¬å–å¯¹æ•°ï¼Œä½†æ˜¯æœ€åŸå§‹çš„æ•°æ®ä¸­æœ‰äº›åŸºå› çš„è®¡æ•°ä¸º0ï¼Œè¿™æ ·åœ¨å–logå€¼çš„æ—¶å€™æ„å‘³ç€âˆ’âˆï¼Œè¿™æ ·æ˜¯ä¸è¡Œçš„ï¼Œæ‰€ä»¥ä¸€èˆ¬ä¼šåŠ ä¸Šä¸€ä¸ªå¸¸æ•°å†å–logï¼Œä¹Ÿå°±æ˜¯log(count + N)ï¼ˆå…¶ä¸­Næ˜¯ä¸€ä¸ªå¸¸æ•°ï¼‰ï¼Œä½†æ˜¯ä¹Ÿæœ‰è¾ƒå¥½çš„æ–¹æ³•æ¥è¿›è¡Œæ ¡æ­£ï¼Œæ¯”å¦‚DEseq2åŒ…è‡ªå¸¦çš„rlogå’Œvstå‡½æ•°ï¼ˆå…¨åä¸ºvariance stabilizing transformationï¼‰ï¼Œå®ƒä»¬æ¶ˆé™¤äº†æ–¹å·®å¯¹å‡å€¼çš„ä¾èµ–ï¼Œå°¤å…¶æ˜¯ä½å‡å€¼æ—¶çš„é«˜log countsçš„å˜å¼‚ã€‚

ä½†æ˜¯åœ¨DESeq2åŒ…ä¸­å®é™…ä¸Šå·²ç»æœ‰äº†å½’ä¸€åŒ–çš„æ–¹æ³•ï¼Œrlogå’Œvstï¼Œåœ¨ä½¿ç”¨çš„éœ€è¦æ ¹æ®æ ·æœ¬é‡çš„å¤šå°‘æ¥é€‰æ‹©æ–¹æ³•ã€‚æ ·æœ¬é‡å°‘äº30çš„è¯ï¼Œé€‰æ‹©rlogï¼Œå¤šäº30çš„è¯ï¼Œå»ºè®®é€‰æ‹©vstã€‚
```
# æ¥ç»­ç€ä¸Šé¢çš„æ„å»ºå¾—åˆ°çš„ddså¯¹è±¡
# DEseq2åŒ…æä¾›äº†ç›¸åº”çš„å‡½æ•°
vsdata <- rlog(dds, blind=FALSE)#rlog è½¬æ¢æ˜¯ä¸€ç§å¸¸ç”¨çš„æ­£æ€åŒ–æ–¹æ³•ï¼Œç”¨äºå‡å°‘è®¡æ•°æ•°æ®çš„ç¦»æ•£æ€§ï¼Œå¹¶å°†æ•°æ®è½¬æ¢ä¸ºè¿‘ä¼¼æ­£æ€åˆ†å¸ƒã€‚è½¬æ¢åçš„æ•°æ®é€‚åˆè¿›è¡Œå·®å¼‚è¡¨è¾¾åˆ†æå’Œå¯è§†åŒ–
#ä½¿ç”¨ rlog å‡½æ•°å¯¹å·®å¼‚è¡¨è¾¾æ•°æ®é›† dds è¿›è¡Œ rlog è½¬æ¢ã€‚blind=FALSE å‚æ•°è¡¨ç¤ºä¸è¿›è¡Œç›²è½¬æ¢ï¼Œè¿™æ ·å‡½æ•°å°†è‡ªåŠ¨é€‰æ‹©æœ€åˆé€‚çš„å‚è€ƒæ ·æœ¬ã€‚
#å˜é‡vsdata

# intgroup åˆ†ç»„
plotPCA(vsdata, intgroup="treatment") + ylim(-10, 10)
```